[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/hien078/applied-mathematics-foundation/blob/master/calculus/04_derivative_applications_optimization/exercises.ipynb)

# Topic 04: Derivative Applications & Optimization — 4-Level Exercise Package & Solutions Manual

---

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import sympy as sp
from scipy import optimize

plt.rcParams.update({
    "figure.figsize": (7.0, 4.0),
    "figure.dpi": 110,
    "axes.grid": True,
    "grid.alpha": 0.3,
    "axes.spines.top": False,
    "axes.spines.right": False,
    "font.size": 11,
})
rng = np.random.default_rng(0)
np.set_printoptions(precision=4, suppress=True)

## L0 — Concept Checks

### Problem L0.1 — Rolle's Theorem Hypotheses Verification
**Source:** Spivak, *Calculus*, Chapter 14  
**Statement.** Consider the function $f(x) = 1 - \lvert x \rvert$ defined on the closed interval $[-1, 1]$. We observe that $f(-1) = 0$ and $f(1) = 0$. Is there any point $c \in (-1, 1)$ such that $f'(c) = 0$? Does this contradict Rolle's Theorem? Explain why or why not from first principles.

#### Intuition
Rolle's Theorem asserts that if a function starts and ends at the same height, it must turn around somewhere, producing a horizontal tangent. However, this turning point must be smooth (differentiable). If the turning point is a sharp corner (v-shape), the tangent line does not exist.

#### Solution
1. **Check Endpoint Values:**  

$$
   f(-1) = 1 - \lvert -1 \rvert = 1 - 1 = 0
$$

$$
   f(1) = 1 - \lvert 1 \rvert = 1 - 1 = 0
$$

   The endpoint equality condition $f(-1) = f(1)$ holds.

2. **Calculate Derivative:**  
   For $x \in (0, 1)$, $f(x) = 1 - x$, so $f'(x) = -1$.  
   For $x \in (-1, 0)$, $f(x) = 1 + x$, so $f'(x) = 1$.  
   At $x = 0$, evaluate left- and right-hand derivatives:

$$
   \lim_{h \to 0^+} \frac{f(0+h) - f(0)}{h} = \lim_{h \to 0^+} \frac{(1 - h) - 1}{h} = -1
$$

$$
   \lim_{h \to 0^-} \frac{f(0+h) - f(0)}{h} = \lim_{h \to 0^-} \frac{(1 + h) - 1}{h} = +1
$$

   Since the left-hand and right-hand limits differ, $f'(0)$ does not exist.

3. **Check Existence of Horizontal Tangent:**  
   $f'(x) = 1 \ne 0$ for $x \in (-1, 0)$ and $f'(x) = -1 \ne 0$ for $x \in (0, 1)$. Thus, there is **no** point $c \in (-1, 1)$ where $f'(c) = 0$.

4. **Reconciliation with Rolle's Theorem:**  
   Rolle's Theorem requires $f$ to be **differentiable on the entire open interval $(-1, 1)$**. Because $f$ is not differentiable at $x = 0 \in (-1, 1)$, hypothesis (2) of Rolle's Theorem is violated. Therefore, Rolle's Theorem does not apply and no contradiction exists.

$$
\boxed{\text{No point } c \text{ exists with } f'(c)=0 \text{. Rolle's Theorem is not violated because } f \text{ is not differentiable at } x=0.}
$$

#### Key takeaway
Differentiability on the *entire* open interval is an indispensable hypothesis of Rolle's Theorem. A single point of non-differentiability can break the guarantee of a horizontal tangent.

**Verification.**

In [2]:
# Rolle's hypotheses: f(x) = 1 - |x| on [-1, 1] has equal endpoint values but a corner at 0.
f = lambda t: 1.0 - np.abs(t)
xs = np.linspace(-1.0, 1.0, 20001)[1:-1]
h = 1e-7
slope = (f(xs + h) - f(xs - h)) / (2 * h)
away = np.abs(xs) > 1e-3
print("f(-1) =", f(-1.0), " f(1) =", f(1.0))
print("smallest |f'| on the interior away from 0:", np.abs(slope[away]).min())
print("one-sided slopes at 0:", (f(h) - f(0.0)) / h, (f(-h) - f(0.0)) / (-h))
assert np.abs(slope[away]).min() > 0.99   # never zero: no Rolle point exists

f(-1) = 0.0  f(1) = 0.0
smallest |f'| on the interior away from 0: 0.9999999994736442
one-sided slopes at 0: -0.9999999994736442 0.9999999994736442


The one-sided slopes at the corner are $-1$ and $+1$: the difference quotient has no limit at $0$, and the derivative is never zero anywhere else, so no Rolle point exists.

### Problem L0.2 — Fermat's Theorem Hypotheses & Domain Boundaries
**Source:** Apostol, *Calculus, Volume I*, Chapter 4  
**Statement.** Consider the linear function $f(x) = x$ defined on the closed domain $[0, 1]$. $f$ attains its absolute maximum at $x = 1$ with $f(1) = 1$ and its absolute minimum at $x = 0$ with $f(0) = 0$. However, $f'(1) = 1 \ne 0$ and $f'(0) = 1 \ne 0$. Explain why this does not contradict Fermat's Theorem on local extrema.

#### Intuition
Fermat's Theorem states that at a local extremum, the tangent line must be flat ($f'(c) = 0$). But this holds ONLY at **interior** points where the function can be approached from both sides. At boundary endpoints, the function stops, so the derivative can be non-zero while still achieving a maximum or minimum.

#### Solution
1. **Fermat's Theorem Statement:**  
   Fermat's Theorem explicitly states: If $f$ has a local extremum at $c$, **and if $c$ is an interior point of the domain**, and if $f'(c)$ exists, then $f'(c) = 0$.

2. **Domain Analysis:**  
   The domain of $f$ is $[0, 1]$. The maximum occurs at $x = 1$ and the minimum at $x = 0$.  
   Neither $x = 0$ nor $x = 1$ is an interior point of $[0, 1]$; they are boundary points.  
   For $x = 1$, left-neighborhood points exist ($x \lt 1$), but no right-neighborhood points exist ($x \gt 1$) within the domain. Thus, the two-sided derivative cannot be defined relative to an open ball around $x = 1$.

3. **Conclusion:**  
   Fermat's Theorem does not apply to boundary endpoints of closed domains. Extrema on boundary points can occur with non-zero derivatives.

$$
\boxed{\text{No contradiction. Fermat's Theorem applies strictly to interior points } c \in (a, b)\text{, not boundary endpoints.}}
$$

#### Key takeaway
When optimizing functions on closed intervals, global extrema can occur at critical points ($f'(c)=0$), non-differentiable points, OR boundary endpoints. Always inspect boundary endpoints separately.

**Verification.**

In [3]:
# Fermat's theorem speaks only about interior points.
fp = lambda t: np.ones_like(np.asarray(t, dtype=float))
xs = np.linspace(0.0, 1.0, 1001)
print("max of f(x) = x on [0,1] at x = 1 (boundary), f'(1) =", fp(1.0))
print("min of f(x) = x on [0,1] at x = 0 (boundary), f'(0) =", fp(0.0))
print("interior critical points:", xs[fp(xs) == 0.0].size)

max of f(x) = x on [0,1] at x = 1 (boundary), f'(1) = 1.0
min of f(x) = x on [0,1] at x = 0 (boundary), f'(0) = 1.0
interior critical points: 0


Both extrema sit on the boundary, where Fermat's theorem says nothing; the interior contains no critical point at all.

### Problem L0.3 — Non-Indeterminate Forms and L'Hôpital Misuse
**Source:** Stewart, *Calculus*, Chapter 4  
**Statement.** Evaluate the limit $\lim_{x \to 0} \frac{\cos x}{x + 1}$. Show what happens if L'Hôpital's Rule is erroneously applied, and explain the correct calculation from first principles.

#### Intuition
L'Hôpital's Rule is a specialized tool for resolving indeterminate ratios ($\frac{0}{0}$ or $\frac{\infty}{\infty}$). Applying it to a determinate limit changes the problem into a completely different (and wrong) limit.

#### Solution
1. **Correct Direct Substitution:**  
   Evaluate numerator and denominator at $x = 0$:

$$
   \text{Numerator: } \lim_{x \to 0} \cos x = \cos(0) = 1
$$

$$
   \text{Denominator: } \lim_{x \to 0} (x + 1) = 0 + 1 = 1
$$

   Since the denominator is non-zero, the quotient limit is straightforward:

$$
   \lim_{x \to 0} \frac{\cos x}{x + 1} = \frac{1}{1} = 1
$$

2. **Erroneous L'Hôpital Application:**  
   If one blindly applies L'Hôpital's Rule by differentiating numerator and denominator:

$$
   \frac{\frac{d}{dx}[\cos x]}{\frac{d}{dx}[x + 1]} = \frac{-\sin x}{1}
$$

   Taking $x \to 0$:

$$
   \lim_{x \to 0} \frac{-\sin x}{1} = \frac{-\sin(0)}{1} = 0
$$

   This gives the incorrect answer $0$.

3. **Why it Failed:**  
   The form was $\frac{1}{1}$, which is determinate, not $\frac{0}{0}$ or $\frac{\infty}{\infty}$. L'Hôpital's Rule requires indeterminacy.

$$
\boxed{\text{Correct Limit} = 1\text{. Erroneous L'Hôpital gives } 0 \text{ because the initial limit was not indeterminate.}}
$$

#### Key takeaway
Always verify that a limit is indeterminate ($\frac{0}{0}$ or $\frac{\pm\infty}{\pm\infty}$) before invoking L'Hôpital's Rule.

**Verification.**

In [4]:
x = sp.symbols("x")
correct = sp.limit(sp.cos(x) / (x + 1), x, 0)
naive = sp.limit(sp.diff(sp.cos(x), x) / sp.diff(x + 1, x), x, 0)
print("correct limit          :", correct)
print("naive L'Hopital answer :", naive)
assert correct == 1 and naive == 0

correct limit          : 1
naive L'Hopital answer : 0


The correct value is $1$; differentiating numerator and denominator of a non-indeterminate quotient produces $0$, an answer with no relation to the limit.

### Problem L0.4 — Zero Second Derivative vs True Inflection Point
**Source:** Spivak, *Calculus*, Chapter 11  
**Statement.** Analyze the functions $f(x) = x^4$ and $g(x) = x^3$ at $x = 0$. Both functions satisfy $f''(0) = 0$ and $g''(0) = 0$. Determine which function possesses a true inflection point at $x = 0$ and explain the fundamental structural difference.

#### Intuition
$f''(c) = 0$ is only a *candidate* condition for an inflection point, analogous to $f'(c) = 0$ being a candidate for a local extremum. A true inflection point requires that the concavity (the sign of $f''$) actually *changes* as $x$ passes through $c$.

#### Solution
1. **Analyze $f(x) = x^4$:**  
   - First derivative: $f'(x) = 4x^3$
   - Second derivative: $f''(x) = 12x^2$
   - At $x = 0$: $f''(0) = 12(0)^2 = 0$.
   - For $x \lt 0$: $f''(x) = 12x^2 \gt 0$ (convex).
   - For $x \gt 0$: $f''(x) = 12x^2 \gt 0$ (convex).
   - Since $f''(x)$ remains positive on both sides of $x=0$, concavity does **not** change. $x = 0$ is a strict local minimum, not an inflection point.

2. **Analyze $g(x) = x^3$:**  
   - First derivative: $g'(x) = 3x^2$
   - Second derivative: $g''(x) = 6x$
   - At $x = 0$: $g''(0) = 6(0) = 0$.
   - For $x \lt 0$: $g''(x) = 6x \lt 0$ (concave).
   - For $x \gt 0$: $g''(x) = 6x \gt 0$ (convex).
   - Since $g''(x)$ changes sign from negative to positive at $x = 0$, concavity changes. Thus $(0, 0)$ is a true inflection point.

$$
\boxed{g(x) = x^3 \text{ has an inflection point at } x=0\text{; } f(x) = x^4 \text{ does not, because } f''(x) \text{ does not change sign.}}
$$

#### Key takeaway
$f''(c) = 0$ is a necessary condition for twice-differentiable inflection points, but the sufficient condition requires a strict sign change of $f''(x)$ across $x = c$.

**Verification.**

In [5]:
for name, fpp in (("x**4", lambda t: 12 * t**2), ("x**3", lambda t: 6 * t)):
    left, right = fpp(-0.1), fpp(0.1)
    print(f"{name}: f''(-0.1) = {left:+8.4f}   f''(+0.1) = {right:+8.4f}   "
          f"sign change: {left * right < 0}")

x**4: f''(-0.1) =  +0.1200   f''(+0.1) =  +0.1200   sign change: False
x**3: f''(-0.1) =  -0.6000   f''(+0.1) =  +0.6000   sign change: True


$f''$ changes sign across $0$ for $x^3$ and does not for $x^4$, which is exactly the criterion for an inflection point.

### Problem L0.5 — Strict Monotonicity with Zero Derivative Points
**Source:** Apostol, *Calculus, Volume I*, Chapter 4  
**Statement.** Let $f(x) = x + \sin x$ on $\mathbb{R}$. Compute $f'(x)$. Is $f'(x) \ge 0$ everywhere? Does $f'(x) = 0$ at isolated points prevent $f(x)$ from being strictly increasing on $\mathbb{R}$? Prove your assertion from first principles.

#### Intuition
A derivative $f'(x) \ge 0$ guarantees that a function is non-decreasing. If $f'(x) = 0$ only at isolated points (and is strictly positive everywhere else), the function cannot remain flat over any interval of positive length, so it remains *strictly* increasing.

#### Solution
1. **Compute Derivative:**  

$$
   f'(x) = \frac{d}{dx}[x + \sin x] = 1 + \cos x
$$

2. **Analyze Derivative Sign:**  
   Since $-1 \le \cos x \le 1$ for all $x \in \mathbb{R}$, we have:

$$
   f'(x) = 1 + \cos x \ge 0 \quad \forall x \in \mathbb{R}
$$

   $f'(x) = 0$ occurs when $\cos x = -1$, which corresponds to $x = (2k + 1)\pi$ for $k \in \mathbb{Z}$. These zero-derivative points are isolated.

3. **First-Principles Strict Monotonicity Proof:**  
   Let $x_1, x_2 \in \mathbb{R}$ with $x_1 \lt x_2$. By the Mean Value Theorem (Theorem 2.4), there exists $c \in (x_1, x_2)$ such that:

$$
   f(x_2) - f(x_1) = f'(c) (x_2 - x_1) = (1 + \cos c)(x_2 - x_1)
$$

   Since $x_2 - x_1 \gt 0$ and $1 + \cos c \ge 0$, we have $f(x_2) - f(x_1) \ge 0$.  
   Could $f(x_1) = f(x_2)$? That would require $1 + \cos c = 0 \implies c = (2k+1)\pi$ for *every* point in $(x_1, x_2)$, meaning $f'(x) = 0$ identically on $(x_1, x_2)$. But $f'(x) = 1 + \cos x$ is zero only at isolated discrete points, not on any open interval.  
   Therefore, $f(x_2) - f(x_1) \gt 0$, so $f(x_1) \lt f(x_2)$.

$$
\boxed{f(x) = x + \sin x \text{ is strictly increasing on } \mathbb{R} \text{ despite } f'(x)=0 \text{ at isolated points } (2k+1)\pi.}
$$

#### Key takeaway
Isolated points where $f'(x) = 0$ do not destroy strict monotonicity, provided $f'$ does not vanish identically on any subinterval.

**Verification.**

In [6]:
xs = np.linspace(-20.0, 20.0, 20001)
f = xs + np.sin(xs)
fp = 1.0 + np.cos(xs)
print("f'(x) >= 0 everywhere sampled :", bool(np.all(fp >= -1e-12)))
print("all forward differences > 0   :", bool(np.all(np.diff(f) > 0)))
print("f' at (2k+1)pi for k = 0,1,2  :", [1 + np.cos((2 * k + 1) * np.pi) for k in range(3)])

f'(x) >= 0 everywhere sampled : True
all forward differences > 0   : True
f' at (2k+1)pi for k = 0,1,2  : [np.float64(0.0), np.float64(0.0), np.float64(0.0)]


The derivative touches $0$ at the isolated points $(2k+1)\pi$ yet every forward difference is strictly positive, so the function is strictly increasing.

### Problem L0.6 — L'Hôpital Failure Due to Oscillatory Derivatives
**Source:** Demidovich, *Problems in Mathematical Analysis*, No. 1388  
**Statement.** Evaluate $\lim_{x \to \infty} \frac{x - \sin x}{x + \sin x}$. Show that applying L'Hôpital's Rule leads to a limit that does not exist, whereas the actual limit exists and can be evaluated by algebraic factorization.

#### Intuition
L'Hôpital's Rule states: IF $\lim \frac{f'}{g'} = L$, THEN $\lim \frac{f}{g} = L$. It does NOT state that if $\lim \frac{f'}{g'}$ fails to exist due to oscillation, $\lim \frac{f}{g}$ must also fail to exist.

#### Solution
1. **Identify Indeterminate Form:**  
   As $x \to \infty$, $x \pm \sin x \to \infty$ because $\lvert\sin x\rvert \le 1$. Form is $\frac{\infty}{\infty}$.

2. **Apply L'Hôpital's Rule:**  

$$
   f'(x) = 1 - \cos x, \quad g'(x) = 1 + \cos x
$$

$$
   \lim_{x \to \infty} \frac{f'(x)}{g'(x)} = \lim_{x \to \infty} \frac{1 - \cos x}{1 + \cos x}
$$

   As $x \to \infty$, $\cos x$ oscillates infinitely between $-1$ and $1$. Thus, the ratio $\frac{1 - \cos x}{1 + \cos x}$ oscillates between $0$ and $\infty$ (undefined at odd multiples of $\pi$). More fundamentally, $g'(x) = 1 + \cos x = 0$ at $x = (2k+1)\pi$ for infinitely many $x$ as $x \to \infty$, so hypothesis (2) of Theorem 3.1 ($g'$ nonzero near the point of interest) already fails: L'Hopital's Rule is not even applicable here, independent of whether the limit of $f'/g'$ exists.

3. **Direct Algebraic Evaluation:**  
   Divide numerator and denominator by $x \gt 0$:

$$
   \frac{x - \sin x}{x + \sin x} = \frac{1 - \frac{\sin x}{x}}{1 + \frac{\sin x}{x}}
$$

   Since $\lvert\sin x\rvert \le 1$, by the Squeeze Theorem, $\lim_{x \to \infty} \frac{\sin x}{x} = 0$. Therefore:

$$
   \lim_{x \to \infty} \frac{1 - \frac{\sin x}{x}}{1 + \frac{\sin x}{x}} = \frac{1 - 0}{1 + 0} = 1
$$

$$
\boxed{\text{Actual Limit} = 1\text{. L'Hôpital fails because } \lim_{x\to\infty}\frac{f'(x)}{g'(x)} \text{ does not exist due to oscillation.}}
$$

#### Key takeaway
The non-existence of $\lim \frac{f'}{g'}$ does not imply the non-existence of $\lim \frac{f}{g}$. When L'Hôpital yields an oscillatory non-existent limit, return to first principles or algebraic bounds.

**Verification.**

In [7]:
xs = np.array([1e2, 1e3, 1e4, 1e5, 1e6])
print("(x - sin x)/(x + sin x) as x grows:", (xs - np.sin(xs)) / (xs + np.sin(xs)))
t = np.linspace(1000.0, 1000.0 + 4 * np.pi, 4001)
ratio = (1 - np.cos(t)) / (1 + np.cos(t))
print(f"f'/g' over one window: min {ratio.min():.4f}  max {ratio.max():.4e}  (no limit)")
assert abs((1e6 - np.sin(1e6)) / (1e6 + np.sin(1e6)) - 1.0) < 1e-5

(x - sin x)/(x + sin x) as x grows: [1.0102 0.9983 1.0001 1.     1.    ]
f'/g' over one window: min 0.0000  max 3.1286e+07  (no limit)


The quotient converges to $1$ while the ratio of derivatives sweeps a whole range on every window, confirming that L'Hopital is inconclusive but the limit exists.

## L1 — Foundations

### Problem L1.1 — Mean Value Theorem Calculation
**Source:** Stewart, *Calculus*, Chapter 4  
**Statement.** Let $f(x) = x^3 - 3x$ on the closed interval $[-1, 2]$. Verify that $f(x)$ satisfies the hypotheses of the Mean Value Theorem on $[-1, 2]$, and find all values $c \in (-1, 2)$ that satisfy the conclusion of the theorem.

#### Intuition
$f(x)$ is a polynomial, hence smooth everywhere. We compute the slope of the secant line across $[-1, 2]$ and solve for points where $f'(c)$ equals this slope.

#### Solution
1. **Verify Hypotheses:**  
   $f(x)$ is a polynomial, so it is continuous on $[-1, 2]$ and differentiable on $(-1, 2)$.

2. **Compute Secant Slope:**  

$$
   f(-1) = (-1)^3 - 3(-1) = -1 + 3 = 2
$$

$$
   f(2) = (2)^3 - 3(2) = 8 - 6 = 2
$$

   Secant slope $m = \frac{f(2) - f(-1)}{2 - (-1)} = \frac{2 - 2}{3} = 0$.

3. **Find $c \in (-1, 2)$ such that $f'(c) = 0$:**  

$$
   f'(x) = 3x^2 - 3
$$

   Set $3c^2 - 3 = 0 \implies c^2 = 1 \implies c = \pm 1$.

4. **Filter Interior Points:**  
   The MVT requires $c \in (-1, 2)$ (open interval).  
   $c = -1$ is an endpoint, so it is excluded.  
   $c = 1 \in (-1, 2)$ is an interior point.

$$
\boxed{c = 1}
$$

#### Key takeaway
Extrema points at domain boundaries must be filtered out; MVT specifically guarantees points in the *open* interval $(a, b)$.

**Verification.**

In [8]:
x = sp.symbols("x")
f = x**3 - 3 * x
a, b = -1, 2
slope = sp.simplify((f.subs(x, b) - f.subs(x, a)) / (b - a))
roots = sp.solve(sp.Eq(sp.diff(f, x), slope), x)
inside = [r for r in roots if a < r < b]
print("secant slope:", slope, "  f'(c) = slope at c in", roots, "  inside (a,b):", inside)
assert inside == [1]

secant slope: 0   f'(c) = slope at c in [-1, 1]   inside (a,b): [1]


Only $c = 1$ lies inside $(-1, 2)$; the other solution $c = -1$ of $f'(c) = 0$ is an endpoint and is therefore excluded.

### Problem L1.2 — Cauchy Mean Value Theorem Application
**Source:** Demidovich, *Problems in Mathematical Analysis*, No. 1240  
**Statement.** Apply Cauchy's Mean Value Theorem to $f(x) = \sin x$ and $g(x) = \cos x$ on the interval $[0, \frac{\pi}{2}]$. Find the exact value of $c \in (0, \frac{\pi}{2})$ guaranteed by the theorem.

#### Intuition
Cauchy MVT relates the ratio of function differences $\frac{f(b)-f(a)}{g(b)-g(a)}$ to the ratio of derivatives $\frac{f'(c)}{g'(c)}$.

#### Solution
1. **Verify Hypotheses:**  
   $f(x) = \sin x$ and $g(x) = \cos x$ are continuous on $[0, \frac{\pi}{2}]$ and differentiable on $(0, \frac{\pi}{2})$.  
   $g'(x) = -\sin x \ne 0$ for all $x \in (0, \frac{\pi}{2})$.

2. **Compute Function Differences:**  

$$
   f(\frac{\pi}{2}) - f(0) = \sin(\frac{\pi}{2}) - \sin(0) = 1 - 0 = 1
$$

$$
   g(\frac{\pi}{2}) - g(0) = \cos(\frac{\pi}{2}) - \cos(0) = 0 - 1 = -1
$$

   Ratio: $\frac{f(\frac{\pi}{2}) - f(0)}{g(\frac{\pi}{2}) - g(0)} = \frac{1}{-1} = -1$.

3. **Set Ratio of Derivatives Equal:**  

$$
   f'(c) = \cos c, \quad g'(c) = -\sin c
$$

$$
   \frac{f'(c)}{g'(c)} = \frac{\cos c}{-\sin c} = -\cot c
$$

4. **Solve for $c$:**  

$$
   -\cot c = -1 \implies \cot c = 1 \implies c = \frac{\pi}{4}
$$

   Since $\frac{\pi}{4} \in (0, \frac{\pi}{2})$, this is the unique valid point.

$$
\boxed{c = \frac{\pi}{4}}
$$

#### Key takeaway
Cauchy's MVT generalizes standard MVT by parametrizing two curves simultaneously, eliminating explicit reliance on the independent variable $x$.

**Verification.**

In [9]:
rhs = (np.sin(np.pi / 2) - np.sin(0.0)) / (np.cos(np.pi / 2) - np.cos(0.0))
c = optimize.brentq(lambda t: np.cos(t) / (-np.sin(t)) - rhs, 1e-8, np.pi / 2 - 1e-8)
print(f"[f(b)-f(a)]/[g(b)-g(a)] = {rhs:.6f}")
print(f"c found by brentq = {c:.12f}   pi/4 = {np.pi / 4:.12f}")
assert abs(c - np.pi / 4) < 1e-10

[f(b)-f(a)]/[g(b)-g(a)] = -1.000000
c found by brentq = 0.785398163397   pi/4 = 0.785398163397


The root-finder lands on $\pi/4$ to ten digits, matching the closed form.

### Problem L1.3 — Indeterminate Form $\frac{0}{0}$ via Repeated L'Hôpital
**Source:** Demidovich, *Problems in Mathematical Analysis*, No. 1321  
**Statement.** Evaluate $\lim_{x \to 0} \frac{x - \sin x}{x^3}$ using L'Hôpital's Rule and verify the result using Taylor series expansion.

#### Intuition
When applying L'Hôpital's Rule yields another $\frac{0}{0}$ indeterminate form, the rule can be applied repeatedly until a non-zero derivative appears in the denominator.

#### Solution
1. **Check Form:** As $x \to 0$, $x - \sin x \to 0$ and $x^3 \to 0$ (Form $\frac{0}{0}$).

2. **First L'Hôpital Pass:**  

$$
   \lim_{x \to 0} \frac{\frac{d}{dx}[x - \sin x]}{\frac{d}{dx}[x^3]} = \lim_{x \to 0} \frac{1 - \cos x}{3x^2} \quad \left(\text{Form }\frac{0}{0}\right)
$$

3. **Second L'Hôpital Pass:**  

$$
   \lim_{x \to 0} \frac{\frac{d}{dx}[1 - \cos x]}{\frac{d}{dx}[3x^2]} = \lim_{x \to 0} \frac{\sin x}{6x} \quad \left(\text{Form }\frac{0}{0}\right)
$$

4. **Third L'Hôpital Pass:**  

$$
   \lim_{x \to 0} \frac{\frac{d}{dx}[\sin x]}{\frac{d}{dx}[6x]} = \lim_{x \to 0} \frac{\cos x}{6} = \frac{1}{6}
$$

5. **Taylor Expansion Verification:**  
   $\sin x = x - \frac{x^3}{6} + O(x^5) \implies x - \sin x = \frac{x^3}{6} - O(x^5)$.  

$$
   \lim_{x \to 0} \frac{\frac{x^3}{6} - O(x^5)}{x^3} = \frac{1}{6}
$$

$$
\boxed{\frac{1}{6}}
$$

#### Key takeaway
Repeated L'Hôpital applications track higher-order terms in the function's Taylor series expansion until the lowest order non-zero polynomial term dominates.

**Verification.**

In [10]:
x = sp.symbols("x")
print("sympy limit  :", sp.limit((x - sp.sin(x)) / x**3, x, 0))
print("Taylor check :", sp.series(x - sp.sin(x), x, 0, 6))
h = np.array([1e-1, 1e-2, 1e-3])
print("numeric ratio:", (h - np.sin(h)) / h**3)

sympy limit  : 1/6


Taylor check : x**3/6 - x**5/120 + O(x**6)
numeric ratio: [0.1666 0.1667 0.1667]


Both the symbolic limit and the Taylor coefficient give $1/6$, and the numeric ratios approach it.

### Problem L1.4 — Indeterminate Form $\frac{\infty}{\infty}$
**Source:** Spivak, *Calculus*, Chapter 11  
**Statement.** Evaluate the limit $\lim_{x \to \infty} \frac{\ln(1 + e^{2x})}{x}$.

#### Intuition
As $x \to \infty$, $e^{2x}$ dominates 1 inside the logarithm, so $\ln(1 + e^{2x}) \approx \ln(e^{2x}) = 2x$. The ratio $\frac{2x}{x}$ should tend to 2.

#### Solution
1. **Check Form:** As $x \to \infty$, $1 + e^{2x} \to \infty \implies \ln(1+e^{2x}) \to \infty$, and $x \to \infty$. Form is $\frac{\infty}{\infty}$.

2. **Apply L'Hôpital's Rule:**  

$$
   \frac{d}{dx}[\ln(1 + e^{2x})] = \frac{1}{1 + e^{2x}} \cdot 2e^{2x} = \frac{2e^{2x}}{1 + e^{2x}}
$$

$$
   \frac{d}{dx}[x] = 1
$$

$$
   \lim_{x \to \infty} \frac{f'(x)}{g'(x)} = \lim_{x \to \infty} \frac{2e^{2x}}{1 + e^{2x}}
$$

3. **Evaluate Secondary Limit:**  
   Divide numerator and denominator by $e^{2x} \gt 0$:

$$
   \lim_{x \to \infty} \frac{2}{e^{-2x} + 1} = \frac{2}{0 + 1} = 2
$$

$$
\boxed{2}
$$

#### Key takeaway
Logarithmic growing functions of exponentials grow linearly; L'Hôpital simplifies composite logarithmic derivatives rapidly.

**Verification.**

In [11]:
x = sp.symbols("x", positive=True)
print("sympy limit:", sp.limit(sp.log(1 + sp.exp(2 * x)) / x, x, sp.oo))
xs = np.array([5.0, 20.0, 100.0, 400.0])
print("numeric    :", np.logaddexp(0.0, 2 * xs) / xs)

sympy limit: 2
numeric    : [2. 2. 2. 2.]


The dominant term $e^{2x}$ forces $\ln(1 + e^{2x}) \sim 2x$, so the ratio tends to $2$.

### Problem L1.5 — Indeterminate Form $0 \cdot \infty$
**Source:** Demidovich, *Problems in Mathematical Analysis*, No. 1345  
**Statement.** Evaluate $\lim_{x \to 0^+} x^2 \ln x$.

#### Intuition
As $x \to 0^+$, $x^2 \to 0$ while $\ln x \to -\infty$. Polynomial decay ($x^2$) is much faster than logarithmic growth/decay ($\lvert\ln x\rvert$), so the product should approach 0.

#### Solution
1. **Rewrite as Fraction ($\frac{\infty}{\infty}$):**  

$$
   x^2 \ln x = \frac{\ln x}{x^{-2}}
$$

   As $x \to 0^+$, $\ln x \to -\infty$ and $x^{-2} \to \infty$. Form is $\frac{\infty}{\infty}$.

2. **Apply L'Hôpital's Rule:**  

$$
   \frac{d}{dx}[\ln x] = \frac{1}{x}
$$

$$
   \frac{d}{dx}[x^{-2}] = -2 x^{-3}
$$

$$
   \lim_{x \to 0^+} \frac{\frac{1}{x}}{-2 x^{-3}} = \lim_{x \to 0^+} \left( \frac{1}{x} \cdot \frac{x^3}{-2} \right) = \lim_{x \to 0^+} \left( -\frac{x^2}{2} \right)
$$

3. **Compute Final Limit:**  

$$
   -\frac{(0)^2}{2} = 0
$$

$$
\boxed{0}
$$

#### Key takeaway
To resolve $0 \cdot \infty$, invert the polynomial term to the denominator as $x^{-p}$; differentiation then reduces the algebraic power of $x$.

**Verification.**

In [12]:
x = sp.symbols("x", positive=True)
print("sympy limit:", sp.limit(x**2 * sp.log(x), x, 0, "+"))
xs = np.array([1e-1, 1e-2, 1e-4, 1e-8])
print("numeric    :", xs**2 * np.log(xs))

sympy limit: 0
numeric    : [-0.023  -0.0005 -0.     -0.    ]


$x^2$ beats $\ln x$: the product collapses to $0$.

### Problem L1.6 — Indeterminate Exponential Form $1^\infty$
**Source:** Stewart, *Calculus*, Chapter 4  
**Statement.** Evaluate $\lim_{x \to 0} (1 + 3x)^{\frac{1}{x}}$.

#### Intuition
This is the canonical definition structure of the natural exponential $e^k$, where $(1 + k/n)^n \to e^k$.

#### Solution
1. **Define Logarithmic Transformation:**  
   Let $y = (1 + 3x)^{\frac{1}{x}}$. Take natural log:

$$
   \ln y = \ln\left((1 + 3x)^{\frac{1}{x}}\right) = \frac{\ln(1 + 3x)}{x}
$$

2. **Evaluate $\lim_{x \to 0} \ln y$ (Form $\frac{0}{0}$):**  
   As $x \to 0$, $\ln(1 + 3x) \to \ln 1 = 0$ and $x \to 0$. Apply L'Hôpital:

$$
   \lim_{x \to 0} \frac{\frac{d}{dx}[\ln(1+3x)]}{\frac{d}{dx}[x]} = \lim_{x \to 0} \frac{\frac{3}{1+3x}}{1} = \frac{3}{1+0} = 3
$$

3. **Exponentiate to Recover Original Limit:**  
   Since $\lim_{x \to 0} \ln y = 3$, by continuity of the exponential function:

$$
   \lim_{x \to 0} y = e^{\lim_{x \to 0} \ln y} = e^3
$$

$$
\boxed{e^3}
$$

#### Key takeaway
Indeterminate exponential forms ($1^\infty, 0^0, \infty^0$) are solved by taking logarithms, converting to $\frac{0}{0}$ or $\frac{\infty}{\infty}$, and exponentiating the final limit.

**Verification.**

In [13]:
x = sp.symbols("x")
val = sp.limit((1 + 3 * x) ** (1 / x), x, 0)
print("sympy limit:", val, "=", float(val))
xs = np.array([1e-2, 1e-4, 1e-6])
print("numeric    :", (1 + 3 * xs) ** (1 / xs), " e**3 =", np.exp(3.0))

sympy limit: exp(3) = 20.085536923187668
numeric    : [19.2186 20.0765 20.0854]  e**3 = 20.085536923187668


The numeric values close in on $e^3 \approx 20.0855$.

### Problem L1.7 — Indeterminate Exponential Form $0^0$
**Source:** Demidovich, *Problems in Mathematical Analysis*, No. 1358  
**Statement.** Evaluate $\lim_{x \to 0^+} x^{\sin x}$.

#### Intuition
Both base $x \to 0$ and exponent $\sin x \to 0$. As $\sin x \approx x$ near zero, $x^x \to 1$.

#### Solution
1. **Logarithmic Transformation:**  
   Let $y = x^{\sin x}$.  

$$
   \ln y = \sin x \ln x = \frac{\ln x}{\csc x} \quad \left(\text{Form }\frac{-\infty}{\infty}\right)
$$

2. **Alternative Algebraic Conversion:**  
   Alternatively, rewrite $\sin x \ln x = \frac{\sin x}{x} \cdot (x \ln x)$.  
   We know $\lim_{x \to 0^+} \frac{\sin x}{x} = 1$.  
   For $\lim_{x \to 0^+} x \ln x$:

$$
   \lim_{x \to 0^+} \frac{\ln x}{x^{-1}} \underbrace{=}_{\text{L'H}} \lim_{x \to 0^+} \frac{\frac{1}{x}}{-\frac{1}{x}^2} = \lim_{x \to 0^+} (-x) = 0
$$

3. **Combine Limits:**  

$$
   \lim_{x \to 0^+} \ln y = \left(\lim_{x \to 0^+} \frac{\sin x}{x}\right) \cdot \left(\lim_{x \to 0^+} x \ln x\right) = 1 \cdot 0 = 0
$$

4. **Exponentiate:**  

$$
   \lim_{x \to 0^+} y = e^0 = 1
$$

$$
\boxed{1}
$$

#### Key takeaway
$0^0$ indeterminate limits resolve to 1 whenever the base is polynomial and the exponent is linear near zero.

**Verification.**

In [14]:
x = sp.symbols("x", positive=True)
print("sympy limit:", sp.limit(x ** sp.sin(x), x, 0, "+"))
xs = np.array([1e-1, 1e-3, 1e-6, 1e-9])
print("numeric    :", xs ** np.sin(xs))

sympy limit: 1
numeric    : [0.7946 0.9931 1.     1.    ]


$\sin x \ln x \to 0$, so the power tends to $e^0 = 1$.

### Problem L1.8 — Complete Monotonicity & Extrema Analysis
**Source:** Demidovich, *Problems in Mathematical Analysis*, No. 1410  
**Statement.** Given $f(x) = 2x^3 - 9x^2 + 12x - 3$, find all critical points, intervals of strict increase and decrease, and classify all local extrema.

#### Intuition
We use the First Derivative Test: $f'(x) \gt 0 \implies$ increasing, $f'(x) \lt 0 \implies$ decreasing. Sign changes across critical points reveal local maxima and minima.

#### Solution
1. **Find Critical Points:**  

$$
   f'(x) = 6x^2 - 18x + 12 = 6(x^2 - 3x + 2) = 6(x - 1)(x - 2)
$$

   Set $f'(x) = 0 \implies x_1 = 1, x_2 = 2$.

2. **Interval Sign Analysis:**  
   - Interval $(-\infty, 1)$: Pick $x = 0 \implies f'(0) = 6(-1)(-2) = 12 \gt 0 \implies$ **Strictly Increasing**.
   - Interval $(1, 2)$: Pick $x = 1.5 \implies f'(1.5) = 6(0.5)(-0.5) = -1.5 \lt 0 \implies$ **Strictly Decreasing**.
   - Interval $(2, \infty)$: Pick $x = 3 \implies f'(3) = 6(2)(1) = 12 \gt 0 \implies$ **Strictly Increasing**.

3. **Classify Extrema:**  
   - At $x = 1$: $f'(x)$ changes from positive to negative $\implies$ **Local Maximum**.  

$$
     f(1) = 2(1)^3 - 9(1)^2 + 12(1) - 3 = 2 - 9 + 12 - 3 = 2
$$

   - At $x = 2$: $f'(x)$ changes from negative to positive $\implies$ **Local Minimum**.  

$$
     f(2) = 2(8) - 9(4) + 12(2) - 3 = 16 - 36 + 24 - 3 = 1
$$

$$
\boxed{\text{Increasing on } (-\infty, 1) \cup (2, \infty)\text{; Decreasing on } (1, 2)\text{. Local Max at } (1, 2)\text{; Local Min at } (2, 1).}
$$

#### Key takeaway
Factoring $f'(x)$ immediately yields the sign chart governing local monotonicity and extremum classification.

**Verification.**

In [15]:
x = sp.symbols("x")
f = 2 * x**3 - 9 * x**2 + 12 * x - 3
crit = sp.solve(sp.diff(f, x), x)
print("critical points:", crit)
for c in crit:
    print(f"  x = {c}:  f = {f.subs(x, c)},  f'' = {sp.diff(f, x, 2).subs(x, c)}")
xs = np.linspace(-1.0, 4.0, 100001)
fp = 6 * xs**2 - 18 * xs + 12
print("f' > 0 on (-1,1):", bool(np.all(fp[(xs > -1) & (xs < 1)] > 0)),
      "  f' < 0 on (1,2):", bool(np.all(fp[(xs > 1.0001) & (xs < 1.9999)] < 0)))

critical points: [1, 2]


  x = 1:  f = 2,  f'' = -6
  x = 2:  f = 1,  f'' = 6
f' > 0 on (-1,1): True   f' < 0 on (1,2): True


$f'' \gt 0$ at $x = 2$ and $f'' \lt 0$ at $x = 1$, confirming the local minimum and maximum, and the derivative signs on the three intervals match the claimed monotonicity.

### Problem L1.9 — Concavity and Inflection Points of Gaussian-Type Curve
**Source:** Stewart, *Calculus*, Chapter 4  
**Statement.** Analyze $f(x) = x e^{-\frac{x^2}{2}}$. Find all intervals of concavity and exact coordinates of all inflection points.

#### Intuition
This is the derivative of a Gaussian envelope. Concavity shifts where the second derivative changes sign ($f''(x) = 0$).

#### Solution
1. **First Derivative:**  

$$
   f'(x) = (1)e^{-\frac{x^2}{2}} + x(-x e^{-\frac{x^2}{2}}) = (1 - x^2) e^{-\frac{x^2}{2}}
$$

2. **Second Derivative:**  

$$
   f''(x) = (-2x)e^{-\frac{x^2}{2}} + (1 - x^2)(-x e^{-\frac{x^2}{2}}) = e^{-\frac{x^2}{2}} \left( -2x - x + x^3 \right) = x(x^2 - 3) e^{-\frac{x^2}{2}}
$$

3. **Find Candidate Inflection Points ($f''(x) = 0$):**  
   Since $e^{-\frac{x^2}{2}} \gt 0$ everywhere, $f''(x) = 0 \implies x(x^2 - 3) = 0$.  
   Candidate roots: $x_1 = -\sqrt{3}, x_2 = 0, x_3 = \sqrt{3}$.

4. **Sign Analysis of $f''(x)$:**  
   - $(-\infty, -\sqrt{3})$: $x \lt 0$ and $x^2-3 \gt 0 \implies f''(x) \lt 0$ (**Concave Down**).
   - $(-\sqrt{3}, 0)$: $x \lt 0$ and $x^2-3 \lt 0 \implies f''(x) \gt 0$ (**Concave Up / Convex**).
   - $(0, \sqrt{3})$: $x \gt 0$ and $x^2-3 \lt 0 \implies f''(x) \lt 0$ (**Concave Down**).
   - $(\sqrt{3}, \infty)$: $x \gt 0$ and $x^2-3 \gt 0 \implies f''(x) \gt 0$ (**Concave Up / Convex**).

5. **Evaluate Inflection Points:**  
   All three candidates change sign, so all three are inflection points:
   - $f(-\sqrt{3}) = -\sqrt{3} e^{-\frac{3}{2}}$
   - $f(0) = 0$
   - $f(\sqrt{3}) = \sqrt{3} e^{-\frac{3}{2}}$

$$
\boxed{\text{Inflection Points at } (0, 0) \text{ and } \left(\pm\sqrt{3}, \pm\sqrt{3} e^{-\frac{3}{2}}\right).}
$$

#### Key takeaway
The inflection points of $x e^{-\frac{x^2}{2}}$ mark the transition boundaries between central exponential decay and outer tail behavior.

**Verification.**

In [16]:
x = sp.symbols("x")
f = x * sp.exp(-x**2 / 2)
fpp = sp.simplify(sp.diff(f, x, 2))
print("f'' =", fpp)
roots = sp.solve(sp.Eq(fpp, 0), x)
print("f'' = 0 at:", roots)
for r in roots:
    print(f"  inflection point ({r}, {sp.simplify(f.subs(x, r))}) "
          f"= ({float(r):+.6f}, {float(f.subs(x, r)):+.6f})")

f'' = x*(x**2 - 3)*exp(-x**2/2)
f'' = 0 at: [0, -sqrt(3), sqrt(3)]
  inflection point (0, 0) = (+0.000000, +0.000000)
  inflection point (-sqrt(3), -sqrt(3)*exp(-3/2)) = (-1.732051, -0.386473)
  inflection point (sqrt(3), sqrt(3)*exp(-3/2)) = (+1.732051, +0.386473)


$f'' = x(x^2 - 3)e^{-x^2/2}$ vanishes at $0$ and $\pm\sqrt{3}$, and $\sqrt{3}e^{-3/2} \approx 0.3866$.

### Problem L1.10 — Global Extrema on Closed Interval
**Source:** Apostol, *Calculus, Volume I*, Chapter 4  
**Statement.** Find the global maximum and global minimum values of $f(x) = x - 2\sin x$ on the closed interval $[0, 2\pi]$.

#### Intuition
Continuous functions on closed intervals attain global extrema either at interior critical points ($f'(x) = 0$) or at interval boundary endpoints.

#### Solution
1. **Find Critical Points in $(0, 2\pi)$:**  

$$
   f'(x) = 1 - 2\cos x
$$

   Set $f'(x) = 0 \implies \cos x = \frac{1}{2}$.  
   In $(0, 2\pi)$, solutions are $x_1 = \frac{\pi}{3}$ and $x_2 = \frac{5\pi}{3}$.

2. **Evaluate $f(x)$ at Candidate Points:**  
   - **Endpoint $x = 0$:**  

$$
     f(0) = 0 - 2\sin(0) = 0
$$

   - **Critical Point $x = \frac{\pi}{3}$:**  

$$
     f\left(\frac{\pi}{3}\right) = \frac{\pi}{3} - 2\sin\left(\frac{\pi}{3}\right) = \frac{\pi}{3} - 2\left(\frac{\sqrt{3}}{2}\right) = \frac{\pi}{3} - \sqrt{3} \approx 1.047 - 1.732 = -0.685
$$

   - **Critical Point $x = \frac{5\pi}{3}$:**  

$$
     f\left(\frac{5\pi}{3}\right) = \frac{5\pi}{3} - 2\sin\left(\frac{5\pi}{3}\right) = \frac{5\pi}{3} - 2\left(-\frac{\sqrt{3}}{2}\right) = \frac{5\pi}{3} + \sqrt{3} \approx 5.236 + 1.732 = 6.968
$$

   - **Endpoint $x = 2\pi$:**  

$$
     f(2\pi) = 2\pi - 2\sin(2\pi) = 2\pi \approx 6.283
$$

3. **Compare Candidates:**  
   - Smallest value: $\frac{\pi}{3} - \sqrt{3} \approx -0.685$.  
   - Largest value: $\frac{5\pi}{3} + \sqrt{3} \approx 6.968$.

$$
\boxed{\text{Global Max } = \frac{5\pi}{3} + \sqrt{3} \quad \text{at } x = \frac{5\pi}{3}; \quad \text{Global Min } = \frac{\pi}{3} - \sqrt{3} \quad \text{at } x = \frac{\pi}{3}}
$$

#### Key takeaway
Never assume endpoints are automatically the extrema; candidate comparison of all critical points and endpoints is mandatory.

**Verification.**

In [17]:
f = lambda t: t - 2 * np.sin(t)
crit = np.array([np.pi / 3, 5 * np.pi / 3])
cand = np.concatenate(([0.0, 2 * np.pi], crit))
vals = f(cand)
print("candidates:", cand)
print("values    :", vals)
print(f"global max = {vals.max():.6f} at x = {cand[vals.argmax()]:.6f} "
      f"(5pi/3 + sqrt(3) = {5 * np.pi / 3 + np.sqrt(3):.6f})")
print(f"global min = {vals.min():.6f} at x = {cand[vals.argmin()]:.6f} "
      f"(pi/3 - sqrt(3) = {np.pi / 3 - np.sqrt(3):.6f})")

candidates: [0.     6.2832 1.0472 5.236 ]
values    : [ 0.      6.2832 -0.6849  6.968 ]
global max = 6.968039 at x = 5.235988 (5pi/3 + sqrt(3) = 6.968039)
global min = -0.684853 at x = 1.047198 (pi/3 - sqrt(3) = -0.684853)


The two interior critical points beat both endpoints, giving the boxed global extrema.

### Problem L1.11 — Newton's Method Iterative Calculation
**Source:** Nocedal & Wright, *Numerical Optimization*, Chapter 2  
**Statement.** Perform two iterations of Newton's method to approximate $\sqrt{5}$ by finding the positive root of $f(x) = x^2 - 5 = 0$, starting with initial guess $x_0 = 2.0$. Express $x_1$ and $x_2$ as exact rational fractions.

#### Intuition
Newton's method iteratively refines an estimate $x_k$ via tangent lines: $x_{k+1} = \frac{1}{2}\left(x_k + \frac{5}{x_k}\right)$ for square roots (the ancient Babylonian algorithm).

#### Solution
1. **Derive Iteration Formula:**  
   $f(x) = x^2 - 5 \implies f'(x) = 2x$.  

$$
   x_{k+1} = x_k - \frac{x_k^2 - 5}{2x_k} = \frac{2x_k^2 - x_k^2 + 5}{2x_k} = \frac{x_k^2 + 5}{2x_k} = \frac{1}{2}\left(x_k + \frac{5}{x_k}\right)
$$

2. **First Iteration ($x_0 = 2 = 2/1$):**  

$$
   x_1 = \frac{1}{2}\left(2 + \frac{5}{2}\right) = \frac{1}{2}\left(\frac{9}{2}\right) = \frac{9}{4} = 2.25
$$

3. **Second Iteration ($x_1 = \frac{9}{4}$):**  

$$
   x_2 = \frac{1}{2}\left(\frac{9}{4} + \frac{5}{\frac{9}{4}}\right) = \frac{1}{2}\left(\frac{9}{4} + \frac{20}{9}\right) = \frac{1}{2}\left(\frac{81 + 80}{36}\right) = \frac{161}{72}
$$

4. **Numerical Precision Check:**  
   $x_2 = \frac{161}{72} \approx 2.236111...$  
   Actual $\sqrt{5} \approx 2.236068...$  
   Error at step 2 is $\lvert x_2 - \sqrt{5} \rvert \approx 0.000043$ (accurate to 4 decimal places after just 2 steps!).

$$
\boxed{x_1 = \frac{9}{4}, \quad x_2 = \frac{161}{72}}
$$

#### Key takeaway
Newton's method quadratic convergence produces exponential growth in decimal precision per step.

**Verification.**

In [18]:
from fractions import Fraction
x = Fraction(2)
for k in (1, 2):
    x = x - (x * x - 5) / (2 * x)
    print(f"x_{k} = {x} = {float(x):.10f}")
assert x == Fraction(161, 72)
print("sqrt(5) =", np.sqrt(5.0), "  |x_2 - sqrt(5)| =", abs(float(x) - np.sqrt(5.0)))

x_1 = 9/4 = 2.2500000000
x_2 = 161/72 = 2.2361111111
sqrt(5) = 2.23606797749979   |x_2 - sqrt(5)| = 4.31336113213554e-05


Exact rational arithmetic reproduces $9/4$ and $161/72$, already accurate to five decimals of $\sqrt{5}$.

### Problem L1.12 — Monotonicity-Based Inequality Proof
**Source:** Spivak, *Calculus*, Chapter 11  
**Statement.** Prove that $\ln(1 + x) \lt x$ for all $x \gt 0$ using derivative monotonicity from first principles.

#### Intuition
To prove $g(x) \lt h(x)$ for $x \gt 0$, construct $f(x) = h(x) - g(x) = x - \ln(1+x)$. If $f(0) = 0$ and $f'(x) \gt 0$ for $x \gt 0$, then $f(x)$ strictly increases away from 0, guaranteeing $f(x) \gt 0$.

#### Solution
1. **Define Auxiliary Function:**  

$$
   f(x) = x - \ln(1 + x) \quad \text{for } x \ge 0
$$

2. **Evaluate at Boundary $x = 0$:**  

$$
   f(0) = 0 - \ln(1 + 0) = 0 - 0 = 0
$$

3. **Compute Derivative for $x \gt 0$:**  

$$
   f'(x) = 1 - \frac{1}{1 + x} = \frac{(1 + x) - 1}{1 + x} = \frac{x}{1 + x}
$$

4. **Analyze Derivative Sign:**  
   For any $x \gt 0$, $x \gt 0$ and $1 + x \gt 0$, so:

$$
   f'(x) = \frac{x}{1 + x} \gt 0 \quad \forall x \gt 0
$$

5. **Apply Monotonicity Theorem:**  
   By Theorem 4.1, $f(x)$ is strictly increasing on $[0, \infty)$.  
   Therefore, for any $x \gt 0$:

$$
   f(x) \gt f(0) \implies x - \ln(1 + x) \gt 0 \implies \ln(1 + x) \lt x
$$

$$
\boxed{\ln(1 + x) \lt x \quad \forall x \gt 0}
$$

#### Key takeaway
Inequalities of the form $g(x) \lt h(x)$ are universally proved by showing $h(x) - g(x)$ has a positive derivative starting from an equality base point.

**Verification.**

In [19]:
xs = np.logspace(-6, 3, 20000)
print("ln(1+x) < x for every sampled x > 0:", bool(np.all(np.log1p(xs) < xs)))
print("largest value of ln(1+x) - x:", (np.log1p(xs) - xs).max())

ln(1+x) < x for every sampled x > 0: True
largest value of ln(1+x) - x: -4.99999666600458e-13


The gap $x - \ln(1+x)$ is strictly positive across nine decades of $x$.

### Problem L1.13 — Geometric Exact Location of MVT Point for Quadratics
**Source:** Stewart, *Calculus*, Chapter 4  
**Statement.** Let $f(x) = Ax^2 + Bx + C$ ($A \ne 0$) be a quadratic polynomial on $[a, b]$. Prove that the point $c \in (a, b)$ guaranteed by the Mean Value Theorem is *always* the exact midpoint of the interval, $c = \frac{a + b}{2}$.

#### Intuition
For any parabola, the slope of the secant line between two points $a$ and $b$ is equal to the instantaneous tangent slope at the exact arithmetic average of $a$ and $b$.

#### Solution
1. **Compute Secant Slope:**  

$$
   f(b) - f(a) = (Ab^2 + Bb + C) - (Aa^2 + Ba + C) = A(b^2 - a^2) + B(b - a)
$$

   Since $b - a \ne 0$, divide by $b - a$:

$$
   \frac{f(b) - f(a)}{b - a} = A(b + a) + B
$$

2. **Compute Derivative & Set Equal:**  
   $f'(x) = 2Ax + B$.  
   Set $f'(c) = \frac{f(b) - f(a)}{b - a}$:

$$
   2Ac + B = A(a + b) + B
$$

3. **Solve for $c$:**  
   Subtract $B$ from both sides:

$$
   2Ac = A(a + b)
$$

   Since $A \ne 0$, divide by $2A$:

$$
   c = \frac{a + b}{2}
$$

   Since $a \lt b$, clearly $a \lt \frac{a+b}{2} \lt b$, so $c \in (a, b)$.

$$
\boxed{c = \frac{a + b}{2}}
$$

#### Key takeaway
For all quadratic functions, the MVT tangent point is precisely the arithmetic mean of the interval endpoints.

**Verification.**

In [20]:
a, b, A, B, C = sp.symbols("a b A B C")
x, c = sp.symbols("x c")
f = A * x**2 + B * x + C
slope = sp.simplify((f.subs(x, b) - f.subs(x, a)) / (b - a))
sol = sp.solve(sp.Eq(sp.diff(f, x).subs(x, c), slope), c)
print("secant slope:", slope)
print("MVT point   :", sp.simplify(sol[0]), " = (a+b)/2 :", sp.simplify(sol[0] - (a + b) / 2) == 0)

secant slope: A*a + A*b + B
MVT point   : a/2 + b/2  = (a+b)/2 : True


Symbolically the MVT point is $(a+b)/2$ for every quadratic, independent of $A$, $B$, $C$.

### Problem L1.14 — Newton's Method Instability Near Zero Derivatives
**Source:** Nocedal & Wright, *Numerical Optimization*, Chapter 2  
**Statement.** Consider applying Newton's method to find a root of $f(x) = x^{\frac{1}{3}}$. Derive the iteration formula $x_{k+1}$ in terms of $x_k$. Show that for any non-zero starting point $x_0 \ne 0$, Newton's method diverges away from the root $r = 0$.

#### Intuition
Newton's method uses local tangent lines. If the function is vertical near the root (so $f'(x) \to \infty$ or $f'(x)$ is very small near turning points), the tangent line shoots the next iterate far away from the root rather than closer to it.

#### Solution
1. **Compute Derivative:**  

$$
   f(x) = x^{\frac{1}{3}} \implies f'(x) = \frac{1}{3} x^{-\frac{2}{3}}
$$

2. **Substitute into Newton's Formula:**  

$$
   x_{k+1} = x_k - \frac{f(x_k)}{f'(x_k)} = x_k - \frac{x_k^{\frac{1}{3}}}{\frac{1}{3} x_k^{-\frac{2}{3}}} = x_k - 3 x_k^{(\frac{1}{3} + \frac{2}{3})} = x_k - 3x_k
$$

$$
   x_{k+1} = -2 x_k
$$

3. **Analyze Iterative Sequence:**  
   Given $x_0 \ne 0$:

$$
   x_1 = -2 x_0
$$

$$
   x_2 = (-2)^2 x_0 = 4 x_0
$$

$$
   x_k = (-2)^k x_0
$$

   As $k \to \infty$, $\lvert x_k \rvert = 2^k \lvert x_0 \rvert \to \infty$.

4. **Conclusion:**  
   The root is $r = 0$. But $\lvert x_k \rvert \to \infty$, so the iterates oscillate with exponentially growing magnitude and diverge to infinity.

$$
\boxed{x_{k+1} = -2x_k\text{. Iterate sequence } x_k = (-2)^k x_0 \text{ diverges to } \infty.}
$$

#### Key takeaway
Newton's method requires $f'(r) \ne 0$ and bounded curvature near the root. Vertical tangents ($f'(x) \to \infty$) or zero slopes ($f'(x) = 0$) cause catastrophic divergence or breakdown.

**Verification.**

In [21]:
x = 1.0
for k in range(6):
    x = x - np.cbrt(x) / ((1.0 / 3.0) * np.cbrt(x) ** (-2))
    print(f"x_{k + 1} = {x:+.4f}")
print("ratio x_{k+1}/x_k is exactly -2, so |x_k| = 2**k grows without bound")

x_1 = -2.0000
x_2 = +4.0000
x_3 = -8.0000
x_4 = +16.0000
x_5 = -32.0000
x_6 = +64.0000
ratio x_{k+1}/x_k is exactly -2, so |x_k| = 2**k grows without bound


Each iterate is exactly $-2$ times the previous one, so the amplitude doubles every step and Newton's method runs away from the root at $0$.

## L2 — Applications (AI/ML and Physics)

### Problem L2.1 — Physics: Snell's Law Derivation from Fermat's Principle
**Source:** Standard Classical Mechanics / Optics  
**Statement.** A light ray travels from point $A=(0, a)$ in medium 1 (speed $v_1$) to point $B=(d, -b)$ in medium 2 (speed $v_2$). The boundary between media is the line $y=0$. Derive Snell's Law $\frac{\sin\theta_1}{v_1} = \frac{\sin\theta_2}{v_2}$ by finding the interface position $x$ that minimizes total travel time.

#### Intuition
Fermat's Principle of Least Time states that light takes the path of minimum travel time. Setting the first derivative of total time $T(x)$ with respect to refraction coordinate $x$ to zero yields Snell's Law.

#### Solution
1. **Formulate Total Travel Time $T(x)$:**  
   Distance in medium 1: $d_1(x) = \sqrt{x^2 + a^2}$. Time $t_1 = \frac{\sqrt{x^2 + a^2}}{v_1}$.  
   Distance in medium 2: $d_2(x) = \sqrt{(d-x)^2 + b^2}$. Time $t_2 = \frac{\sqrt{(d-x)^2 + b^2}}{v_2}$.  

$$
   T(x) = \frac{\sqrt{x^2 + a^2}}{v_1} + \frac{\sqrt{(d-x)^2 + b^2}}{v_2}
$$

2. **Compute First Derivative $T'(x)$:**  

$$
   T'(x) = \frac{x}{v_1 \sqrt{x^2 + a^2}} - \frac{d - x}{v_2 \sqrt{(d-x)^2 + b^2}}
$$

3. **Relate to Angles of Incidence & Refraction:**  
   From the right triangles at the interface:

$$
   \sin\theta_1 = \frac{x}{\sqrt{x^2 + a^2}}, \quad \sin\theta_2 = \frac{d - x}{\sqrt{(d-x)^2 + b^2}}
$$

4. **Set $T'(x^{\ast}) = 0$ for Optimality:**  

$$
   \frac{\sin\theta_1}{v_1} - \frac{\sin\theta_2}{v_2} = 0 \implies \frac{\sin\theta_1}{v_1} = \frac{\sin\theta_2}{v_2}
$$

5. **Verify Strict Convexity ($T''(x) \gt 0$):**  

$$
   T''(x) = \frac{a^2}{v_1 (x^2 + a^2)^{\frac{3}{2}}} + \frac{b^2}{v_2 ((d-x)^2 + b^2)^{\frac{3}{2}}} \gt 0
$$

   Since $T''(x) \gt 0$ everywhere, $x^{\ast}$ is the unique global minimum time path.

$$
\boxed{\frac{\sin\theta_1}{\sin\theta_2} = \frac{v_1}{v_2}}
$$

#### Key takeaway
Physical laws governing wave refraction (Snell's Law) emerge as the unique global minimum of a 1D convex travel time optimization problem.

**Verification.**

In [22]:
a, b, d, v1, v2 = 1.0, 1.5, 2.0, 1.0, 0.7
T = lambda t: np.sqrt(t**2 + a**2) / v1 + np.sqrt((d - t) ** 2 + b**2) / v2
res = optimize.minimize_scalar(T, bounds=(0.0, d), method="bounded",
                               options={"xatol": 1e-12})
xs = res.x
s1 = xs / np.sqrt(xs**2 + a**2)
s2 = (d - xs) / np.sqrt((d - xs) ** 2 + b**2)
print(f"time-minimising crossing point x* = {xs:.10f}")
print(f"sin(theta1)/v1 = {s1 / v1:.12f}   sin(theta2)/v2 = {s2 / v2:.12f}")
print(f"sin(theta1)/sin(theta2) = {s1 / s2:.10f}   v1/v2 = {v1 / v2:.10f}")
assert abs(s1 / v1 - s2 / v2) < 1e-6   # limited by sqrt(eps) near a smooth minimum

time-minimising crossing point x* = 1.0945325767
sin(theta1)/v1 = 0.738269103724   sin(theta2)/v2 = 0.738269129381
sin(theta1)/sin(theta2) = 1.4285713789   v1/v2 = 1.4285714286


The numerically optimal crossing point satisfies $\sin\theta_1 / v_1 = \sin\theta_2 / v_2$ to nine digits, which is Snell's law.

### Problem L2.2 — Physics: Maximum Power Transfer Theorem
**Source:** Electrical Engineering / Physics  
**Statement.** A DC voltage source with EMF $V$ and internal resistance $r$ is connected to a variable load resistor $R \gt 0$. The electrical power dissipated in the load resistor is $P(R) = \frac{V^2 R}{(R + r)^2}$. Find the load resistance $R^{\ast}$ that maximizes power transfer, and compute the maximum power $P_{\text{max}}$.

#### Intuition
If $R \to 0$, current is high but voltage across load is zero ($P \to 0$). If $R \to \infty$, voltage is high but current is zero ($P \to 0$). The peak power must occur at a finite intermediate resistance $R^{\ast}$.

#### Solution
1. **Compute Derivative $P'(R)$:**  
   Using the quotient rule on $P(R) = V^2 \cdot \frac{R}{(R+r)^2}$:

$$
   P'(R) = V^2 \frac{(1)(R+r)^2 - R \cdot 2(R+r)}{(R+r)^4} = V^2 \frac{(R+r) - 2R}{(R+r)^3} = V^2 \frac{r - R}{(R+r)^3}
$$

2. **Find Critical Point:**  
   Set $P'(R) = 0 \implies r - R = 0 \implies R^{\ast} = r$.

3. **Verify Local Maximum via First Derivative Test:**  
   - For $0 \lt R \lt r$: $r - R \gt 0 \implies P'(R) \gt 0$ (increasing).
   - For $R \gt r$: $r - R \lt 0 \implies P'(R) \lt 0$ (decreasing).  
   Hence $R^{\ast} = r$ is the global maximum.

4. **Compute Maximum Power $P_{\text{max}}$:**  

$$
   P_{\text{max}} = P(r) = \frac{V^2 r}{(r + r)^2} = \frac{V^2 r}{4r^2} = \frac{V^2}{4r}
$$

$$
\boxed{R^{\ast} = r, \quad P_{\text{max}} = \frac{V^2}{4r}}
$$

#### Key takeaway
Maximum power transfer occurs when load impedance matches internal source impedance ($R=r$), achieving 50% electrical efficiency.

**Verification.**

In [23]:
V, r = 12.0, 3.0
P = lambda R: V**2 * R / (R + r) ** 2
res = optimize.minimize_scalar(lambda R: -P(R), bounds=(1e-6, 100.0), method="bounded",
                               options={"xatol": 1e-12})
print(f"numerical argmax R* = {res.x:.10f}   theory r = {r}")
print(f"P(R*) = {P(res.x):.10f}   theory V^2/(4r) = {V**2 / (4 * r):.10f}")
assert abs(res.x - r) < 1e-6 and abs(P(res.x) - V**2 / (4 * r)) < 1e-9

numerical argmax R* = 3.0000000445   theory r = 3.0
P(R*) = 12.0000000000   theory V^2/(4r) = 12.0000000000


Maximum power transfer occurs at $R^{\star} = r$, with peak power $V^2/(4r)$, matching the closed form.

### Problem L2.3 — Physics: Optimal Projectile Range on Inclined Plane
**Source:** Classical Mechanics (Marion & Thornton)  
**Statement.** A projectile is launched from ground level with initial speed $v_0$ at angle $\theta$ relative to the horizontal. The horizontal range is $R(\theta) = \frac{v_0^2 \sin(2\theta)}{g}$ for $\theta \in [0, \frac{\pi}{2}]$. Find the optimal launch angle $\theta^{\ast}$ that maximizes range on flat ground.

#### Intuition
Range combines horizontal velocity component $v_0 \cos\theta$ and time of flight $t = \frac{2v_0\sin\theta}{g}$. Trade-off between speed and flight time reaches peak at $\theta = 45^\circ$.

#### Solution
1. **Compute Derivative $R'(\theta)$:**  

$$
   R'(\theta) = \frac{v_0^2}{g} \frac{d}{d\theta}[\sin(2\theta)] = \frac{2 v_0^2}{g} \cos(2\theta)
$$

2. **Find Critical Point:**  
   Set $R'(\theta) = 0 \implies \cos(2\theta) = 0$.  
   For $\theta \in (0, \frac{\pi}{2})$, $2\theta \in (0, \pi)$, so $2\theta = \frac{\pi}{2} \implies \theta^{\ast} = \frac{\pi}{4} = 45^\circ$.

3. **Second Derivative Test:**  

$$
   R''(\theta) = -\frac{4 v_0^2}{g} \sin(2\theta)
$$

   At $\theta^{\ast} = \frac{\pi}{4}$:

$$
   R''(\frac{\pi}{4}) = -\frac{4 v_0^2}{g} \sin(\frac{\pi}{2}) = -\frac{4 v_0^2}{g} \lt 0
$$

   Confirms a strict maximum.

4. **Compute Maximum Range:**  

$$
   R_{\text{max}} = R(\frac{\pi}{4}) = \frac{v_0^2 \sin(\frac{\pi}{2})}{g} = \frac{v_0^2}{g}
$$

$$
\boxed{\theta^{\ast} = 45^\circ \left(\frac{\pi}{4}\right), \quad R_{\text{max}} = \frac{v_0^2}{g}}
$$

#### Key takeaway
$45^\circ$ perfectly balances the competing linear factors of horizontal propulsion and vertical flight duration.

**Verification.**

In [24]:
v0, g = 20.0, 9.81
R = lambda th: v0**2 * np.sin(2 * th) / g
res = optimize.minimize_scalar(lambda th: -R(th), bounds=(0.0, np.pi / 2),
                               method="bounded", options={"xatol": 1e-12})
print(f"numerical argmax theta* = {np.degrees(res.x):.8f} deg   theory 45 deg")
print(f"R(theta*) = {R(res.x):.8f} m   theory v0^2/g = {v0**2 / g:.8f} m")
assert abs(np.degrees(res.x) - 45.0) < 1e-6

numerical argmax theta* = 45.00000000 deg   theory 45 deg
R(theta*) = 40.77471967 m   theory v0^2/g = 40.77471967 m


The optimizer returns $45^\circ$ and $R_{\max} = v_0^2/g$.

### Problem L2.4 — AI/ML: Binary Cross-Entropy Loss Strict Convexity
**Source:** Goodfellow et al., *Deep Learning*, Chapter 4  
**Statement.** Let $y \in \{0, 1\}$ be a target binary label, $x \ne 0$ a real feature, and $w \in \mathbb{R}$ a scalar weight. The model prediction is $p = \sigma(w x) = \frac{1}{1 + e^{-w x}}$. The Binary Cross-Entropy (BCE) loss is:

$$
L(w) = -y \ln(\sigma(w x)) - (1-y) \ln(1 - \sigma(w x))
$$

Prove that $L(w)$ is strictly convex with respect to $w$ for all $w \in \mathbb{R}$.

#### Intuition
A loss function with positive second derivative ($L''(w) \gt 0$) is strictly convex, guaranteeing that gradient descent cannot get stuck in local sub-optimal minima.

#### Solution
1. **Recall Sigmoid Derivative Identity:**  
   Let $z = w x$. $\sigma(z) = \frac{1}{1 + e^{-z}}$.  

$$
   \frac{d\sigma}{dz} = \sigma(z)(1 - \sigma(z))
$$

   By chain rule, $\frac{d\sigma}{dw} = x \sigma(w x)(1 - \sigma(w x))$.

2. **Compute First Derivative $L'(w)$:**  

$$
   L'(w) = -y \frac{x \sigma(wx)(1-\sigma(wx))}{\sigma(wx)} - (1-y) \frac{-x \sigma(wx)(1-\sigma(wx))}{1-\sigma(wx)}
$$

$$
   L'(w) = -y x (1 - \sigma(w x)) + (1-y) x \sigma(w x) = x \left[ \sigma(w x) - y \right]
$$

3. **Compute Second Derivative $L''(w)$:**  

$$
   L''(w) = \frac{d}{dw}\left( x [\sigma(w x) - y] \right) = x \cdot x \sigma(w x)(1 - \sigma(w x)) = x^2 \sigma(w x)(1 - \sigma(w x))
$$

4. **Establish Strict Inequality:**  
   - Since $x \ne 0$, $x^2 \gt 0$.
   - For any real $z$, $\sigma(z) \in (0, 1)$, so $\sigma(w x) \gt 0$ and $1 - \sigma(w x) \gt 0$.
   - Product of strictly positive terms: $L''(w) \gt 0$ for all $w \in \mathbb{R}$.

5. **Conclusion:**  
   By Theorem 4.2, $L(w)$ is strictly convex on $\mathbb{R}$.

$$
\boxed{L''(w) = x^2 \sigma(wx)(1-\sigma(wx)) \gt 0 \implies L(w) \text{ is strictly convex on } \mathbb{R}.}
$$

#### Key takeaway
The combination of sigmoid activations with cross-entropy loss eliminates non-convex loss plateaus, yielding global convergence guarantees.

**Verification.**

In [25]:
sig = lambda z: 1.0 / (1.0 + np.exp(-z))
xf, y = 1.7, 1.0
L = lambda w: -y * np.log(sig(w * xf)) - (1 - y) * np.log(1 - sig(w * xf))
ws = np.linspace(-6.0, 6.0, 4001)
h = 1e-4
num_second = (L(ws + h) - 2 * L(ws) + L(ws - h)) / h**2
analytic = xf**2 * sig(ws * xf) * (1 - sig(ws * xf))
print("max |numeric L'' - analytic L''|:", np.abs(num_second - analytic).max())
print("analytic L'' strictly positive everywhere sampled:", bool(np.all(analytic > 0)))

max |numeric L'' - analytic L''|: 5.094476199458747e-07
analytic L'' strictly positive everywhere sampled: True


The finite-difference second derivative agrees with $x^2 \sigma(1-\sigma)$ to within the discretization error, and it never touches zero.

### Problem L2.5 — AI/ML: Ridge Regression 1D Closed-Form Solution
**Source:** Hastie et al., *The Elements of Statistical Learning*  
**Statement.** Consider a 1D L2-regularized least-squares objective (Ridge Regression):

$$
L(w) = \frac{1}{2} \sum_{i=1}^N (y_i - w x_i)^2 + \frac{\lambda}{2} w^2 \quad (\lambda \gt 0)
$$

Derive the analytical global minimizer $w^{\ast}$ from first principles and prove it is unique.

#### Intuition
L2 regularization adds a quadratic penalty $\lambda w^2$ to control model complexity. The objective remains quadratic, yielding a unique closed-form solution.

#### Solution
1. **Expand Loss Function:**  

$$
   L(w) = \frac{1}{2} \sum_{i=1}^N \left( y_i^2 - 2 w x_i y_i + w^2 x_i^2 \right) + \frac{\lambda}{2} w^2
$$

$$
   L(w) = \frac{1}{2} \left( \sum_{i=1}^N x_i^2 + \lambda \right) w^2 - \left( \sum_{i=1}^N x_i y_i \right) w + \frac{1}{2} \sum_{i=1}^N y_i^2
$$

2. **Compute First Derivative $L'(w)$:**  

$$
   L'(w) = \left( \sum_{i=1}^N x_i^2 + \lambda \right) w - \sum_{i=1}^N x_i y_i
$$

3. **Set $L'(w^{\ast}) = 0$ to Solve for $w^{\ast}$:**  

$$
   \left( \sum_{i=1}^N x_i^2 + \lambda \right) w^{\ast} = \sum_{i=1}^N x_i y_i \implies w^{\ast} = \frac{\sum_{i=1}^N x_i y_i}{\sum_{i=1}^N x_i^2 + \lambda}
$$

4. **Verify Strict Convexity ($L''(w) \gt 0$):**  

$$
   L''(w) = \sum_{i=1}^N x_i^2 + \lambda
$$

   Since $\sum x_i^2 \ge 0$ and $\lambda \gt 0$, $L''(w) \ge \lambda \gt 0$.  
   Thus $L(w)$ is strictly convex and $w^{\ast}$ is the unique global minimizer.

$$
\boxed{w^{\ast} = \frac{\sum_{i=1}^N x_i y_i}{\sum_{i=1}^N x_i^2 + \lambda}}
$$

#### Key takeaway
L2 regularization acts as a positive shift $\lambda$ in the denominator, preventing division by zero even when features $x_i = 0$.

**Verification.**

In [26]:
xd = rng.normal(size=40)
yd = 2.3 * xd + 0.4 * rng.normal(size=40)
lam = 0.75
w_closed = (xd @ yd) / (xd @ xd + lam)
L = lambda w: 0.5 * np.sum((yd - w * xd) ** 2) + 0.5 * lam * w**2
res = optimize.minimize_scalar(L, bracket=(-5.0, 5.0))
print(f"closed form w* = {w_closed:.12f}")
print(f"numerical  w* = {res.x:.12f}")
print(f"L''        = {xd @ xd + lam:.6f} > 0, so the minimiser is unique")
assert abs(w_closed - res.x) < 1e-6

closed form w* = 2.189783151743
numerical  w* = 2.189783151743
L''        = 25.678690 > 0, so the minimiser is unique


The closed-form ridge solution and the numerical minimizer agree; the curvature $\sum x_i^2 + \lambda$ is strictly positive, so the minimizer is unique.

### Problem L2.6 — AI/ML: Huber Loss Smoothness & Convexity
**Source:** Nocedal & Wright, *Numerical Optimization*, Chapter 3  
**Statement.** The Huber loss function combines $L_2$ square loss near origin with $L_1$ linear loss in tails:

$$
H_\delta(e) = \begin{cases} \frac{1}{2} e^2 & \text{if } \lvert e \rvert \le \delta \\ \delta \lvert e \rvert - \frac{1}{2} \delta^2 & \text{if } \lvert e \rvert \gt \delta \end{cases}
$$

Prove that $H_\delta(e)$ is continuously differentiable ($C^1$) everywhere and convex on $\mathbb{R}$.

#### Intuition
Standard $L_1$ loss $\lvert e \rvert$ has a non-differentiable sharp corner at $e=0$. Huber loss smooths this corner with a parabola $\frac{1}{2}e^2$, retaining robustness to outliers without losing differentiability.

#### Solution
1. **Check Derivative for $\lvert e \rvert \lt \delta$:**  
   $H'_\delta(e) = e$.

2. **Check Derivative for $\lvert e \rvert \gt \delta$:**  
   - For $e \gt \delta$: $H_\delta(e) = \delta e - \frac{1}{2}\delta^2 \implies H'_\delta(e) = \delta$.
   - For $e \lt -\delta$: $H_\delta(e) = -\delta e - \frac{1}{2}\delta^2 \implies H'_\delta(e) = -\delta$.

3. **Verify Differentiability at Boundaries $e = \pm \delta$:**  
   - At $e = \delta$:  
     Left derivative: $\lim_{e \to \delta^-} e = \delta$.  
     Right derivative: $\lim_{e \to \delta^+} \delta = \delta$.  
     They match! $H'_\delta(\delta) = \delta$.
   - At $e = -\delta$:  
     Left derivative: $\lim_{e \to -\delta^-} (-\delta) = -\delta$.  
     Right derivative: $\lim_{e \to -\delta^+} e = -\delta$.  
     They match! $H'_\delta(-\delta) = -\delta$.

4. **Express Full Derivative:**  

$$
   H'_\delta(e) = \begin{cases} -\delta & \text{if } e \lt -\delta \\ e & \text{if } \lvert e \rvert \le \delta \\ \delta & \text{if } e \gt \delta \end{cases}
$$

   $H'_\delta(e)$ is continuous everywhere on $\mathbb{R}$, so $H_\delta \in C^1(\mathbb{R})$.

5. **Verify Monotonicity of $H'_\delta(e)$ for Convexity:**  
   $H'_\delta(e)$ is non-decreasing on $(-\infty, \infty)$ (it stays flat at $-\delta$, ramps linearly up to $\delta$, and stays flat at $\delta$). By Theorem 4.2 (via the equivalence $f$ convex $\iff$ $f'$ non-decreasing proved there), a continuous function with a non-decreasing derivative is **convex**.

$$
\boxed{H'_\delta(e) \text{ is continuous everywhere and non-decreasing } \implies H_\delta \in C^1(\mathbb{R}) \text{ and is convex.}}
$$

#### Key takeaway
Huber loss provides smooth derivative transitions at $e = \pm\delta$, enabling robust optimization without gradient discontinuities.

**Verification.**

In [27]:
delta = 1.0
H = lambda e: np.where(np.abs(e) <= delta, 0.5 * e**2, delta * np.abs(e) - 0.5 * delta**2)
Hp = lambda e: np.clip(e, -delta, delta)
es = np.linspace(-3.0, 3.0, 60001)
h = 1e-6
num = (H(es + h) - H(es - h)) / (2 * h)
print("max |numeric H' - clip(e)|      :", np.abs(num - Hp(es)).max())
print("H' is non-decreasing            :", bool(np.all(np.diff(Hp(es)) >= -1e-15)))
lam = 0.37
e1, e2 = -2.4, 1.8
print("chord test H(lam e1+(1-lam)e2) <= lam H(e1)+(1-lam)H(e2):",
      bool(H(np.array([lam * e1 + (1 - lam) * e2]))[0]
           <= lam * H(np.array([e1]))[0] + (1 - lam) * H(np.array([e2]))[0]))

max |numeric H' - clip(e)|      : 2.5002122505668467e-07
H' is non-decreasing            : True
chord test H(lam e1+(1-lam)e2) <= lam H(e1)+(1-lam)H(e2): True


The numerical derivative equals the clipped identity everywhere, that clipped function is non-decreasing, and the chord test passes: $H_\delta$ is $C^1$ and convex.

### Problem L2.7 — AI/ML: 1D Exact Line Search for Quadratic Loss
**Source:** Nocedal & Wright, *Numerical Optimization*, Chapter 3  
**Statement.** In optimization, given current point $x_k$ and search direction $d \ne 0$, exact line search computes step size $\alpha^{\ast} = \arg\min_{\alpha \gt 0} \phi(\alpha)$ where $\phi(\alpha) = f(x_k + \alpha d)$. Derive $\alpha^{\ast}$ for quadratic objective $f(x) = \frac{1}{2} a x^2 - b x$ ($a \gt 0$).

#### Intuition
Along a 1D slice, a 2D/multi-D quadratic function reduces to a simple parabola in $\alpha$. The minimum occurs where $\phi'(\alpha) = 0$.

#### Solution
1. **Construct 1D Function $\phi(\alpha)$:**  

$$
   \phi(\alpha) = f(x_k + \alpha d) = \frac{1}{2} a (x_k + \alpha d)^2 - b (x_k + \alpha d)
$$

$$
   \phi(\alpha) = \frac{1}{2} a \left( x_k^2 + 2\alpha x_k d + \alpha^2 d^2 \right) - b x_k - \alpha b d
$$

2. **Compute Derivative $\phi'(\alpha)$ via Chain Rule:**  

$$
   \phi'(\alpha) = \nabla f(x_k + \alpha d) \cdot d = \left[ a(x_k + \alpha d) - b \right] d
$$

$$
   \phi'(\alpha) = (a x_k - b)d + a d^2 \alpha
$$

   Note that $g_k = f'(x_k) = a x_k - b$, so $\phi'(\alpha) = g_k d + a d^2 \alpha$.

3. **Set $\phi'(\alpha^{\ast}) = 0$:**  

$$
   g_k d + a d^2 \alpha^{\ast} = 0 \implies \alpha^{\ast} = -\frac{g_k d}{a d^2}
$$

4. **Special Case: Steepest Descent ($d = -g_k$):**  

$$
   \alpha^{\ast} = -\frac{g_k (-g_k)}{a (-g_k)^2} = \frac{g_k^2}{a g_k^2} = \frac{1}{a}
$$

$$
\boxed{\alpha^{\ast} = -\frac{f'(x_k) d}{a d^2}}
$$

#### Key takeaway
For quadratic objectives, exact line search step size depends directly on the directional derivative divided by curvature $a d^2$.

**Verification.**

In [28]:
a_q, b_q, xk, d = 2.5, 1.1, -0.8, 1.4
f = lambda t: 0.5 * a_q * t**2 - b_q * t
phi = lambda al: f(xk + al * d)
fp_xk = a_q * xk - b_q
alpha_star = -fp_xk * d / (a_q * d**2)
res = optimize.minimize_scalar(phi, bracket=(-5.0, 5.0))
print(f"closed form alpha* = {alpha_star:.12f}")
print(f"numerical  alpha* = {res.x:.12f}")
print(f"one exact line search lands on the global minimiser: "
      f"{xk + alpha_star * d:.12f} vs b/a = {b_q / a_q:.12f}")
assert abs(alpha_star - res.x) < 1e-8

closed form alpha* = 0.885714285714
numerical  alpha* = 0.885714285714
one exact line search lands on the global minimiser: 0.440000000000 vs b/a = 0.440000000000


The closed-form step and the numerical exact line search agree, and one exact step reaches the global minimizer of the quadratic.

### Problem L2.8 — AI/ML: Logistic Activation Peak Gradient & Inflection Point
**Source:** Goodfellow et al., *Deep Learning*, Chapter 6  
**Statement.** Let $f(x) = \sigma(x) = \frac{1}{1 + e^{-x}}$ be the standard logistic activation function. Find the point $x^{\ast}$ where the gradient $f'(x)$ attains its absolute maximum, compute $f'(x^{\ast})$, and prove that $(x^{\ast}, f(x^{\ast}))$ is an inflection point of $\sigma(x)$.

#### Intuition
The sigmoid curve transitions from flat ($0$) to flat ($1$). Its steepest growth (maximum derivative) must occur at the symmetric midpoint $x=0$, which is also its inflection point.

#### Solution
1. **First Derivative:**  

$$
   f'(x) = \sigma(x)(1 - \sigma(x)) = \frac{e^{-x}}{(1 + e^{-x})^2}
$$

2. **Second Derivative:**  
   Using product rule on $f'(x) = \sigma(x) - \sigma(x)^2$:

$$
   f''(x) = \sigma'(x) - 2\sigma(x)\sigma'(x) = \sigma'(x)(1 - 2\sigma(x)) = \sigma(x)(1 - \sigma(x))(1 - 2\sigma(x))
$$

3. **Find Critical Points of $f'(x)$ ($f''(x) = 0$):**  
   Since $\sigma(x) \in (0, 1)$ for all $x \in \mathbb{R}$, $\sigma(x) \ne 0$ and $1 - \sigma(x) \ne 0$.  
   Set $1 - 2\sigma(x) = 0 \implies \sigma(x) = \frac{1}{2}$.  

$$
   \frac{1}{1 + e^{-x}} = \frac{1}{2} \implies 1 + e^{-x} = 2 \implies e^{-x} = 1 \implies x^{\ast} = 0
$$

4. **Compute Maximum Derivative Value:**  

$$
   f'(0) = \sigma(0)(1 - \sigma(0)) = \left(\frac{1}{2}\right)\left(1 - \frac{1}{2}\right) = \frac{1}{4}
$$

5. **Verify Inflection Point Sign Change:**  
   - For $x \lt 0$: $\sigma(x) \lt \frac{1}{2} \implies 1 - 2\sigma(x) \gt 0 \implies f''(x) \gt 0$ (convex up).
   - For $x \gt 0$: $\sigma(x) \gt \frac{1}{2} \implies 1 - 2\sigma(x) \lt 0 \implies f''(x) \lt 0$ (concave down).  
   $f''(x)$ changes sign across $x = 0$, confirming an inflection point at $(0, \frac{1}{2})$.

$$
\boxed{x^{\ast} = 0, \quad f'(0) = \frac{1}{4} \quad (\text{Maximum Gradient and Inflection Point at } (0, \frac{1}{2}))}
$$

#### Key takeaway
The maximum slope of the standard logistic sigmoid is $\frac{1}{4}$ at $x=0$, explaining why deep networks with sigmoid activations suffer from vanishing gradients ($f'(x) \le \frac{1}{4}$).

**Verification.**

In [29]:
sig = lambda z: 1.0 / (1.0 + np.exp(-z))
zs = np.linspace(-8.0, 8.0, 160001)
sp1 = sig(zs) * (1 - sig(zs))
print(f"max of sigma' = {sp1.max():.12f} at z = {zs[sp1.argmax()]:+.6f}   theory 0.25 at z = 0")
print(f"sigma(0) = {sig(0.0):.4f}, so the inflection point is (0, 1/2)")
h = 1e-4
sig_pp = (sig(zs + h) - 2 * sig(zs) + sig(zs - h)) / h**2
print("sigma'' changes sign at z = 0:", sig_pp[zs < -0.1].max() > 0 > sig_pp[zs > 0.1].min())

max of sigma' = 0.250000000000 at z = +0.000000   theory 0.25 at z = 0
sigma(0) = 0.5000, so the inflection point is (0, 1/2)


sigma'' changes sign at z = 0: True


The peak slope is exactly $1/4$ at $z = 0$, and $\sigma''$ changes sign there, so $(0, 1/2)$ is the inflection point.

### Problem L2.9 — AI/ML: Softplus Function Curvature & Smooth ReLU Bounds
**Source:** Goodfellow et al., *Deep Learning*, Chapter 6  
**Statement.** The Softplus activation function is $f(x) = \ln(1 + e^x)$. Prove that:
1. $f(x) \gt \max(0, x)$ for all $x \in \mathbb{R}$.
2. $f'(x) \in (0, 1)$ for all $x \in \mathbb{R}$.
3. $f''(x) \gt 0$ for all $x \in \mathbb{R}$ (strictly convex).

#### Intuition
Softplus is a smooth, strictly convex approximation of the non-differentiable ReLU function $\max(0, x)$.

#### Solution
1. **Prove $f(x) \gt \max(0, x)$:**  
   - For any $x$: $e^x \gt 0 \implies 1 + e^x \gt 1 \implies \ln(1 + e^x) \gt \ln 1 = 0$.
   - Rewrite $f(x) = \ln(e^x(e^{-x} + 1)) = \ln(e^x) + \ln(1 + e^{-x}) = x + \ln(1 + e^{-x})$.
   - Since $e^{-x} \gt 0$, $\ln(1 + e^{-x}) \gt 0$, so $f(x) \gt x$.
   - Combining both bounds: $f(x) \gt 0$ and $f(x) \gt x \implies f(x) \gt \max(0, x)$.

2. **Compute Derivative $f'(x)$:**  

$$
   f'(x) = \frac{e^x}{1 + e^x} = \frac{1}{1 + e^{-x}} = \sigma(x)
$$

   Since $\sigma(x)$ is the logistic sigmoid, $f'(x) \in (0, 1)$ for all $x \in \mathbb{R}$.

3. **Compute Second Derivative $f''(x)$:**  

$$
   f''(x) = \frac{d}{dx}[\sigma(x)] = \sigma(x)(1 - \sigma(x)) = \frac{e^x}{(1 + e^x)^2}
$$

   Since $e^x \gt 0$ and $(1 + e^x)^2 \gt 0$, we have $f''(x) \gt 0$ everywhere. By Theorem 4.2, $f(x)$ is strictly convex.

$$
\boxed{f(x) \gt \max(0, x), \quad f'(x) = \sigma(x) \in (0, 1), \quad f''(x) \gt 0 \text{ (Strictly Convex)}}
$$

#### Key takeaway
Softplus antiderivative is precisely the logistic sigmoid function $\sigma(x)$; its strict convexity ensures smooth optimization without zero-gradient dead zones.

**Verification.**

In [30]:
xs = np.linspace(-8.0, 8.0, 160001)
softplus = np.logaddexp(0.0, xs)
sig = 1.0 / (1.0 + np.exp(-xs))
h = 1e-4
d1 = (np.logaddexp(0.0, xs + h) - np.logaddexp(0.0, xs - h)) / (2 * h)
d2 = (np.logaddexp(0.0, xs + h) - 2 * softplus + np.logaddexp(0.0, xs - h)) / h**2
print("softplus > max(0, x) everywhere:", bool(np.all(softplus > np.maximum(0.0, xs))))
print("max |numeric f' - sigmoid|     :", np.abs(d1 - sig).max())
print("f'' > 0 everywhere sampled     :", bool(np.all(d2 > 0)))

softplus > max(0, x) everywhere: True
max |numeric f' - sigmoid|     : 1.6161216809251755e-10
f'' > 0 everywhere sampled     : True


Softplus dominates the ReLU, its derivative is the sigmoid, and its second derivative is strictly positive.

### Problem L2.10 — AI/ML: 1D Newton-Raphson Update for Logistic Loss
**Source:** Hastie et al., *The Elements of Statistical Learning*  
**Statement.** Consider optimizing weight $w \in \mathbb{R}$ for a single sample with loss $L(w) = \ln(1 + e^{-w y x})$ ($y \in \{-1, +1\}$). Derive the explicit 1D Newton iteration step $w_{k+1} = w_k - \frac{L'(w_k)}{L''(w_k)}$.

#### Intuition
Newton's method scales first-order gradients by inverse curvature $\frac{1}{L''(w_k)}$, speeding up convergence near optimal weights.

#### Solution
1. **Compute First Derivative $L'(w)$:**  
   Let $z = -w y x$.  

$$
   L'(w) = \frac{1}{1 + e^{-w y x}} \cdot \left( -y x e^{-w y x} \right) = \frac{-y x}{e^{w y x} + 1} = -y x \sigma(-w y x)
$$

2. **Compute Second Derivative $L''(w)$:**  

$$
   L''(w) = \frac{d}{dw}\left( \frac{-y x}{1 + e^{w y x}} \right) = -y x \cdot \frac{-y x e^{w y x}}{(1 + e^{w y x})^2} = x^2 y^2 \frac{e^{w y x}}{(1 + e^{w y x})^2}
$$

   Since $y \in \{-1, +1\}$, $y^2 = 1$.  

$$
   L''(w) = x^2 \sigma(w y x)(1 - \sigma(w y x)) = x^2 \sigma(w y x) \sigma(-w y x)
$$

3. **Construct Newton Iteration:**  

$$
   w_{k+1} = w_k - \frac{-y x \sigma(-w_k y x)}{x^2 \sigma(w_k y x) \sigma(-w_k y x)} = w_k + \frac{y x}{x^2 \sigma(w_k y x)} = w_k + \frac{y}{x \sigma(w_k y x)}
$$

$$
\boxed{w_{k+1} = w_k + \frac{y}{x \, \sigma(w_k y x)}}
$$

#### Key takeaway
Newton steps for logistic loss automatically boost updates when prediction confidence $\sigma(w_k y x)$ is low.

**Verification.**

In [31]:
sig = lambda z: 1.0 / (1.0 + np.exp(-z))
xf, y, w0 = 1.3, -1.0, 0.6
L = lambda w: np.logaddexp(0.0, -w * y * xf)
h = 1e-5
Lp = (L(w0 + h) - L(w0 - h)) / (2 * h)
Lpp = (L(w0 + h) - 2 * L(w0) + L(w0 - h)) / h**2
newton_numeric = w0 - Lp / Lpp
newton_closed = w0 + y / (xf * sig(w0 * y * xf))
print(f"finite-difference Newton step: {newton_numeric:.9f}")
print(f"closed-form   Newton step   : {newton_closed:.9f}")
assert abs(newton_numeric - newton_closed) < 1e-5

finite-difference Newton step: -1.847290422
closed-form   Newton step   : -1.847286358


The finite-difference Newton step matches the closed-form update to five digits.

### Problem L2.11 — Physics: Cylindrical Container Minimum Surface Area
**Source:** Stewart, *Calculus*, Chapter 4  
**Statement.** A closed cylindrical can (with top and bottom lids) must enclose a fixed volume $V_0$. Find the ratio of height $h$ to radius $r$ that minimizes the total surface area $A$.

#### Intuition
For a given volume, squishy tall cylinders have large side areas, while short fat cylinders have large top/bottom lid areas. The optimal proportion balances side and lid contributions.

#### Solution
1. **Formulate Constraints & Objective:**  
   Volume constraint: $V = \pi r^2 h = V_0 \implies h = \frac{V_0}{\pi r^2}$.  
   Surface Area objective: $A(r, h) = 2\pi r^2 + 2\pi r h$.  
   Substitute $h(r)$:

$$
   A(r) = 2\pi r^2 + 2\pi r \left( \frac{V_0}{\pi r^2} \right) = 2\pi r^2 + \frac{2V_0}{r}
$$

2. **Compute Derivative $A'(r)$:**  

$$
   A'(r) = 4\pi r - \frac{2V_0}{r^2}
$$

3. **Set $A'(r^{\ast}) = 0$:**  

$$
   4\pi r - \frac{2V_0}{r^2} = 0 \implies 4\pi r^3 = 2V_0 \implies V_0 = 2\pi r^3
$$

4. **Find Height-to-Radius Ratio:**  
   Substitute $V_0 = \pi r^2 h$ into $V_0 = 2\pi r^3$:

$$
   \pi r^2 h = 2\pi r^3 \implies h = 2r \implies \frac{h}{r} = 2
$$

5. **Second Derivative Test:**  

$$
   A''(r) = 4\pi + \frac{4V_0}{r^3} \gt 0 \quad \text{for } r \gt 0
$$

   Confirms a unique global minimum.

$$
\boxed{h = 2r \quad \left(\text{Height equals Diameter, } \frac{h}{r} = 2\right)}
$$

#### Key takeaway
An optimal closed cylinder inscribed in a square cross-section ($h = 2r$) minimizes surface area for a fixed volume.

**Verification.**

In [32]:
Vol = 1.0
A = lambda r: 2 * np.pi * r**2 + 2 * Vol / r
res = optimize.minimize_scalar(A, bounds=(1e-3, 5.0), method="bounded",
                               options={"xatol": 1e-13})
r_star = res.x
h_star = Vol / (np.pi * r_star**2)
print(f"r* = {r_star:.10f}   h* = {h_star:.10f}   h*/r* = {h_star / r_star:.10f}")
print(f"theory r* = (V/(2 pi))**(1/3) = {(Vol / (2 * np.pi)) ** (1 / 3):.10f}")
assert abs(h_star / r_star - 2.0) < 1e-6

r* = 0.5419260624   h* = 1.0838521712   h*/r* = 2.0000000856
theory r* = (V/(2 pi))**(1/3) = 0.5419260701


The minimizing can has $h = 2r$ exactly.

### Problem L2.12 — Physics: Falling Body Terminal Velocity Optimization
**Source:** Classical Mechanics  
**Statement.** A falling mass $m$ under constant gravity $g$ experiences quadratic air resistance force $F_d(v) = b v^2$ ($b \gt 0$). The net force equation is $m \frac{dv}{dt} = m g - b v^2$. Find the terminal velocity $v_T = \lim_{t \to \infty} v(t)$ by finding the equilibrium point where net force acceleration vanishes, and analyze its stability using derivative signs.

#### Intuition
Terminal velocity occurs when drag balances weight ($m g = b v^2$), causing acceleration $\frac{dv}{dt}$ to drop to zero.

#### Solution
1. **Set Acceleration $\frac{dv}{dt} = 0$:**  

$$
   m g - b v_T^2 = 0 \implies b v_T^2 = m g \implies v_T = \sqrt{\frac{m g}{b}}
$$

2. **Analyze Dynamic Stability via Derivative:**  
   Let $f(v) = \frac{dv}{dt} = g - \frac{b}{m} v^2$.  
   Derivative with respect to $v$:

$$
   f'(v) = -\frac{2b}{m} v
$$

   At equilibrium $v_T = \sqrt{\frac{mg}{b}} \gt 0$:

$$
   f'(v_T) = -\frac{2b}{m} \sqrt{\frac{mg}{b}} = -\frac{2\sqrt{bgm}}{m} \lt 0
$$

3. **Interpret Derivative Sign:**  
   Since $f'(v_T) \lt 0$, any small perturbation $\delta v$:
   - If $v \gt v_T$, $\frac{dv}{dt} \lt 0 \implies v$ decreases back to $v_T$.
   - If $v \lt v_T$, $\frac{dv}{dt} \gt 0 \implies v$ increases back to $v_T$.  
   Thus $v_T$ is a **dynamically stable fixed point**.

$$
\boxed{v_T = \sqrt{\frac{mg}{b}}, \quad f'(v_T) \lt 0 \text{ (Dynamically Stable Fixed Point)}}
$$

#### Key takeaway
Equilibrium points of dynamical systems with negative derivative slopes ($f'(v^{\ast}) \lt 0$) are asymptotically stable attractors.

**Verification.**

In [33]:
m, g, b = 2.0, 9.81, 0.35
v = lambda t: m * g - b * t**2          # net force; v' = F/m
v_T = optimize.brentq(v, 0.0, 100.0)
print(f"terminal speed v_T = {v_T:.10f}   theory sqrt(mg/b) = {np.sqrt(m * g / b):.10f}")
h = 1e-6
print(f"d/dv (F/m) at v_T = {(v(v_T + h) - v(v_T - h)) / (2 * h) / m:+.6f} < 0, "
      f"so v_T is an asymptotically stable fixed point")
assert abs(v_T - np.sqrt(m * g / b)) < 1e-9

terminal speed v_T = 7.4871318178   theory sqrt(mg/b) = 7.4871318178
d/dv (F/m) at v_T = -2.620496 < 0, so v_T is an asymptotically stable fixed point


The net force vanishes at $\sqrt{mg/b}$ and its slope there is negative, so the terminal speed is an attracting fixed point.

## L3 — Challenge Proofs

### Problem L3.1 — Putnam 1990 A1 — Polynomial Derivative Inequality
**Source:** Classical analysis result via logarithmic differentiation; see Polya & Szego, *Problems and Theorems in Analysis I*, Part I.  
**Statement.** Let $f(x)$ be a polynomial with real coefficients having $n$ distinct real roots. Prove that $(f'(x))^2 - f(x) f''(x) \gt 0$ for all real $x$ strictly between the smallest and largest roots.

#### Intuition
Consider the logarithmic derivative $\frac{d}{dx}\left[\frac{f'(x)}{f(x)}\right]$. Using the quotient rule, this derivative yields $\frac{f''(x) f(x) - (f'(x))^2}{(f(x))^2}$. If we express $f(x)$ as a product of linear root factors, this derivative can be shown to be strictly negative.

#### Solution
1. **Factor $f(x)$ into Roots:**  
   Since $f(x)$ has $n$ distinct real roots $r_1 \lt r_2 \lt \dots \lt r_n$, factor $f(x)$:

$$
   f(x) = A \prod_{i=1}^n (x - r_i) \quad (A \ne 0)
$$

2. **Compute Logarithmic Derivative:**  
   For any $x \ne r_i$:

$$
   \ln \lvert f(x) \rvert = \ln \lvert A \rvert + \sum_{i=1}^n \ln \lvert x - r_i \rvert
$$

   Differentiate with respect to $x$:

$$
   \frac{f'(x)}{f(x)} = \sum_{i=1}^n \frac{1}{x - r_i}
$$

3. **Differentiate Again:**  

$$
   \frac{d}{dx} \left[ \frac{f'(x)}{f(x)} \right] = \frac{f''(x) f(x) - (f'(x))^2}{(f(x))^2}
$$

   On the right-hand side, differentiate each term $\frac{1}{x - r_i}$:

$$
   \frac{d}{dx} \left[ \sum_{i=1}^n \frac{1}{x - r_i} \right] = -\sum_{i=1}^n \frac{1}{(x - r_i)^2}
$$

4. **Equate and Simplify:**  

$$
   \frac{f''(x) f(x) - (f'(x))^2}{(f(x))^2} = -\sum_{i=1}^n \frac{1}{(x - r_i)^2}
$$

   Multiply both sides by $-(f(x))^2 \lt 0$:

$$
   (f'(x))^2 - f(x) f''(x) = (f(x))^2 \sum_{i=1}^n \frac{1}{(x - r_i)^2}
$$

5. **Establish Strict Positivity:**  
   Since $(f(x))^2 \gt 0$ for all $x \ne r_i$ and each $\frac{1}{(x - r_i)^2} \gt 0$, the right-hand side is a sum of strictly positive terms.  
   What if $x = r_k$ for some root?  
   If $x = r_k$, then $f(r_k) = 0$. Since all $n$ roots are distinct, $f'(r_k) \ne 0$. Thus $(f'(r_k))^2 - 0 = (f'(r_k))^2 \gt 0$.  
   Therefore, the inequality holds strictly for all $x \in \mathbb{R}$.

$$
\boxed{(f'(x))^2 - f(x) f''(x) = (f(x))^2 \sum_{i=1}^n \frac{1}{(x - r_i)^2} \quad \forall x \ne r_i}
$$

$$
\boxed{(f'(x))^2 - f(x) f''(x) \gt 0 \quad \forall x \in \mathbb{R}}
$$

#### Key takeaway
Logarithmic differentiation transforms product structures into additive inverse-square sums, guaranteeing strict positive curvature bounds.

In [34]:
roots = np.array([-2.0, 0.5, 3.0])

def f_poly(x):
    return np.prod(x - roots)

def fprime_poly(x):
    return sum(np.prod(np.delete(x - roots, i)) for i in range(len(roots)))

def fdoubleprime_poly(x):
    total = 0.0
    for i in range(len(roots)):
        for j in range(len(roots)):
            if j == i:
                continue
            total += np.prod(np.delete(x - roots, [i, j]))
    return total

xs = np.linspace(-5, 6, 2000)
xs = xs[np.all(np.abs(xs[:, None] - roots[None, :]) > 1e-6, axis=1)]
lhs = np.array([fprime_poly(x) ** 2 - f_poly(x) * fdoubleprime_poly(x) for x in xs])
print("min of (f')^2 - f*f'' over sampled x (excluding roots):", lhs.min())
print("strictly positive everywhere sampled:", np.all(lhs > 0))

min of (f')^2 - f*f'' over sampled x (excluding roots): 39.06250000017191
strictly positive everywhere sampled: True


The numerically evaluated expression stays strictly positive away from the roots, confirming the
boxed inequality.

### Problem L3.2 — Generalized 3-Function Cauchy MVT Determinant Form
**Source:** Demidovich / Kaczor & Nowak, *Problems in Mathematical Analysis*  
**Statement.** Let $f, g, h$ be continuous on $[a, b]$ and differentiable on $(a, b)$. Define the matrix determinant function $D(x)$:

$$
D(x) = \begin{vmatrix} f(x) & g(x) & h(x) \\ f(a) & g(a) & h(a) \\ f(b) & g(b) & h(b) \end{vmatrix}
$$

Prove that there exists at least one point $c \in (a, b)$ such that $D'(c) = 0$, which yields the generalized determinant Mean Value Theorem:

$$
\begin{vmatrix} f'(c) & g'(c) & h'(c) \\ f(a) & g(a) & h(a) \\ f(b) & g(b) & h(b) \end{vmatrix} = 0
$$

#### Intuition
This is the determinant formulation of Rolle's Theorem. Evaluating $D(x)$ at boundary endpoints $x=a$ and $x=b$ creates identical rows, causing $D(a) = 0$ and $D(b) = 0$.

#### Solution
1. **Evaluate $D(a)$:**  

$$
   D(a) = \begin{vmatrix} f(a) & g(a) & h(a) \\ f(a) & g(a) & h(a) \\ f(b) & g(b) & h(b) \end{vmatrix} = 0 \quad (\text{Row 1 equals Row 2})
$$

2. **Evaluate $D(b)$:**  

$$
   D(b) = \begin{vmatrix} f(b) & g(b) & h(b) \\ f(a) & g(a) & h(a) \\ f(b) & g(b) & h(b) \end{vmatrix} = 0 \quad (\text{Row 1 equals Row 3})
$$

3. **Verify Differentiability:**  
   $D(x)$ is a linear combination of $f(x), g(x), h(x)$ (by cofactor expansion along Row 1). Since $f, g, h$ are continuous on $[a, b]$ and differentiable on $(a, b)$, $D(x)$ is continuous on $[a, b]$ and differentiable on $(a, b)$.

4. **Apply Rolle's Theorem to $D(x)$:**  
   Since $D(a) = D(b) = 0$, by Rolle's Theorem (Theorem 2.3), there exists $c \in (a, b)$ such that $D'(c) = 0$.

5. **Compute $D'(c)$:**  
   By linearity of differentiation across Row 1:

$$
   D'(c) = \begin{vmatrix} f'(c) & g'(c) & h'(c) \\ f(a) & g(a) & h(a) \\ f(b) & g(b) & h(b) \end{vmatrix} = 0
$$

$$
\boxed{\begin{vmatrix} f'(c) & g'(c) & h'(c) \\ f(a) & g(a) & h(a) \\ f(b) & g(b) & h(b) \end{vmatrix} = 0 \quad \text{for some } c \in (a, b)}
$$

#### Key takeaway
Determinant formulations provide a coordinate-free, unified proof of Cauchy MVT and standard MVT in higher dimensions.

**Verification.**

In [35]:
f = lambda t: np.sin(t)
g = lambda t: t**2 + 1.0
hh = lambda t: np.exp(0.3 * t)
a, b = 0.4, 2.1
def D(t):
    return np.linalg.det(np.array([[f(t), g(t), hh(t)],
                                   [f(a), g(a), hh(a)],
                                   [f(b), g(b), hh(b)]]))
print(f"D(a) = {D(a):.3e}   D(b) = {D(b):.3e}   (both zero: repeated rows)")
eps = 1e-6
Dp = lambda t: (D(t + eps) - D(t - eps)) / (2 * eps)
c = optimize.brentq(Dp, a + 1e-3, b - 1e-3)
print(f"Rolle point c = {c:.10f} in (a, b), D'(c) = {Dp(c):.3e}")

D(a) = 0.000e+00   D(b) = -4.909e-16   (both zero: repeated rows)
Rolle point c = 1.2840526384 in (a, b), D'(c) = 0.000e+00


$D$ vanishes at both endpoints because two rows coincide, and Rolle produces an interior $c$ with $D'(c) = 0$, which is the determinant identity.

### Problem L3.3 — Superlinear Convergence Order of Secant Method
**Source:** Nocedal & Wright, *Numerical Optimization*, Chapter 3  
**Statement.** The Secant Method replaces $f'(x_k)$ in Newton's method with a finite difference approximation:

$$
x_{k+1} = x_k - f(x_k) \frac{x_k - x_{k-1}}{f(x_k) - f(x_{k-1})}
$$

Derive the error relation $e_{k+1} \approx C e_k e_{k-1}$ near a simple root $r$ ($f(r)=0, f'(r)\ne 0$) and prove that the order of convergence is the Golden Ratio $\phi = \frac{1 + \sqrt{5}}{2} \approx 1.618$.

#### Intuition
Secant method avoids computing exact derivatives $f'(x)$. Using two points introduces memory ($e_{k-1}$), dropping the convergence order from $2.0$ (Newton) to $\phi \approx 1.618$ (superlinear).

#### Solution
1. **Express Secant Step in Terms of Errors:**  
   Let $e_k = x_k - r$.  

$$
   x_{k+1} = x_k - \frac{f(x_k)}{\frac{f(x_k) - f(x_{k-1})}{x_k - x_{k-1}}} \implies e_{k+1} = e_k - \frac{f(x_k)}{f[x_k, x_{k-1}]}
$$

2. **Taylor Series Expansion of Divided Difference:**  
   Using Taylor expansions $f(x_k) = f'(r) e_k + \frac{1}{2} f''(r) e_k^2 + O(e_k^3)$:

$$
   f[x_k, x_{k-1}] = \frac{f(x_k) - f(x_{k-1})}{e_k - e_{k-1}} = f'(r) + \frac{1}{2} f''(r) (e_k + e_{k-1}) + O(e^2)
$$

3. **Derive Error Relationship:**  

$$
   e_{k+1} = e_k - \frac{f'(r) e_k + \frac{1}{2} f''(r) e_k^2}{f'(r) + \frac{1}{2} f''(r) (e_k + e_{k-1})} = \frac{\frac{1}{2} f''(r)}{f'(r)} e_k e_{k-1} + O(e_k^2 e_{k-1})
$$

$$
   e_{k+1} \approx C e_k e_{k-1} \quad \text{where } C = \frac{f''(r)}{2 f'(r)}
$$

4. **Determine Order of Convergence $p$:**  
   Assume $\lvert e_{k+1} \rvert \approx K \lvert e_k \rvert^p$. Then $\lvert e_k \rvert \approx K \lvert e_{k-1} \rvert^p \implies \lvert e_{k-1} \rvert \approx K^{-1/p} \lvert e_k \rvert^{1/p}$.  
   Substitute into error equation:

$$
   K \lvert e_k \rvert^p = C \lvert e_k \rvert \left( K^{-1/p} \lvert e_k \rvert^{1/p} \right) = C K^{-1/p} \lvert e_k \rvert^{1 + 1/p}
$$

5. **Equate Powers of $\lvert e_k \rvert$:**  

$$
   p = 1 + \frac{1}{p} \implies p^2 - p - 1 = 0
$$

   Applying the quadratic formula for positive root $p$:

$$
   p = \frac{1 + \sqrt{1 - 4(1)(-1)}}{2} = \frac{1 + \sqrt{5}}{2} = \phi \approx 1.618
$$

$$
\boxed{p = \frac{1 + \sqrt{5}}{2} \approx 1.618 \quad (\text{Golden Ratio Convergence Order})}
$$

#### Key takeaway
Derivative-free secant iterations achieve superlinear convergence order $\phi \approx 1.618$ per step without evaluating second derivatives.

**Verification.**

In [36]:
f = lambda t: t**3 - 2 * t - 5
r = optimize.brentq(f, 2.0, 3.0, xtol=1e-15, rtol=8.9e-16)
x0, x1 = 2.0, 3.0
errs = [abs(x0 - r), abs(x1 - r)]
for _ in range(6):
    x2 = x1 - f(x1) * (x1 - x0) / (f(x1) - f(x0))
    x0, x1 = x1, x2
    errs.append(abs(x1 - r))
errs = np.array(errs)
usable = errs[(errs < 1e-2) & (errs > 1e-13)]
order = np.log(usable[1:]) / np.log(usable[:-1])
print("secant errors :", errs)
print("measured order:", order)
print("golden ratio  :", (1 + np.sqrt(5)) / 2)

secant errors : [0.0946 0.9054 0.0357 0.0133 0.0003 0.     0.     0.    ]
measured order: [1.5958 1.6705]
golden ratio  : 1.618033988749895


The measured convergence order settles near $1.618$, the golden ratio.

### Problem L3.4 — Cambridge Tripos — Derivative Bound via MVT
**Source:** Cambridge Mathematical Tripos, Part IA  
**Statement.** Let $f: \mathbb{R} \to \mathbb{R}$ be twice differentiable on $\mathbb{R}$. Suppose $\lvert f(x) \rvert \le A$ and $\lvert f''(x) \rvert \le B$ for all $x \in \mathbb{R}$ for positive constants $A, B \gt 0$. Prove that:

$$
\lvert f'(x) \rvert \le 2 \sqrt{A B} \quad \forall x \in \mathbb{R}
$$

#### Intuition
If a function is bounded ($\lvert f \rvert \le A$) and cannot curve too fast ($\lvert f'' \rvert \le B$), its derivative (slope) cannot be arbitrarily steep. We use Taylor series with remainder of step size $h \gt 0$ to minimize the combined bound.

#### Solution
1. **Apply Taylor's Theorem with Remainder:**  
   For any $x \in \mathbb{R}$ and any step $h \gt 0$:

$$
   f(x + h) = f(x) + h f'(x) + \frac{h^2}{2} f''(\xi_1) \quad (\xi_1 \in (x, x+h))
$$

$$
   f(x - h) = f(x) - h f'(x) + \frac{h^2}{2} f''(\xi_2) \quad (\xi_2 \in (x-h, x))
$$

2. **Subtract Equations to Isolate $f'(x)$:**  

$$
   f(x + h) - f(x - h) = 2h f'(x) + \frac{h^2}{2} \left[ f''(\xi_1) - f''(\xi_2) \right]
$$

$$
   2h f'(x) = f(x + h) - f(x - h) - \frac{h^2}{2} \left[ f''(\xi_1) - f''(\xi_2) \right]
$$

   Divide by $2h$:

$$
   f'(x) = \frac{f(x+h) - f(x-h)}{2h} - \frac{h}{4} \left[ f''(\xi_1) - f''(\xi_2) \right]
$$

3. **Apply Triangle Inequality & Absolute Bounds:**  
   Since $\lvert f(x) \rvert \le A$ and $\lvert f''(x) \rvert \le B$:

$$
   \lvert f(x+h) - f(x-h) \rvert \le \lvert f(x+h) \rvert + \lvert f(x-h) \rvert \le A + A = 2A
$$

$$
   \lvert f''(\xi_1) - f''(\xi_2) \rvert \le \lvert f''(\xi_1) \rvert + \lvert f''(\xi_2) \rvert \le B + B = 2B
$$

   Thus:

$$
   \lvert f'(x) \rvert \le \frac{2A}{2h} + \frac{h}{4} (2B) = \frac{A}{h} + \frac{B h}{2}
$$

4. **Optimize Choice of Free Parameter $h \gt 0$:**  
   Define $g(h) = \frac{A}{h} + \frac{B h}{2}$.  
   Find $h^{\ast}$ minimizing $g(h)$:

$$
   g'(h) = -\frac{A}{h^2} + \frac{B}{2} = 0 \implies h^2 = \frac{2A}{B} \implies h^{\ast} = \sqrt{\frac{2A}{B}}
$$

5. **Substitute Optimal $h^{\ast}$:**  

$$
   \lvert f'(x) \rvert \le \frac{A}{\sqrt{\frac{2A}{B}}} + \frac{B}{2} \sqrt{\frac{2A}{B}} = \sqrt{\frac{AB}{2}} + \sqrt{\frac{AB}{2}} = 2 \sqrt{\frac{AB}{2}} = \sqrt{2AB} \le 2\sqrt{AB}
$$

   More precisely: $\lvert f'(x) \rvert \le \sqrt{2AB}$. Since $\sqrt{2AB} \lt 2\sqrt{AB}$, the requested upper bound $\lvert f'(x) \rvert \le 2\sqrt{AB}$ holds strictly.

$$
\boxed{\lvert f'(x) \rvert \le \sqrt{2AB} \le 2\sqrt{AB}}
$$

#### Key takeaway
Coupling Taylor expansion step sizes $h$ with optimization provides tight derivative bounds from function and second derivative limits.

**Verification.**

In [37]:
A, B = 1.0, 1.0
f = np.sin                        # |f| <= 1 = A and |f''| <= 1 = B on all of R
xs = np.linspace(-10.0, 10.0, 200001)
print(f"sup |f'| measured = {np.abs(np.cos(xs)).max():.10f}")
print(f"bound sqrt(2AB)   = {np.sqrt(2 * A * B):.10f}   (the bound is not tight here)")
print(f"weaker 2 sqrt(AB) = {2 * np.sqrt(A * B):.10f}")
assert np.abs(np.cos(xs)).max() <= np.sqrt(2 * A * B) + 1e-12

sup |f'| measured = 1.0000000000
bound sqrt(2AB)   = 1.4142135624   (the bound is not tight here)
weaker 2 sqrt(AB) = 2.0000000000


The measured supremum of $\lvert f' \rvert$ respects the Landau bound $\sqrt{2AB}$.

### Problem L3.5 — First-Principles Calculus Proof of Jensen's Inequality
**Source:** Polya & Szegö, *Problems and Theorems in Analysis I*  
**Statement.** Let $f: I \to \mathbb{R}$ be a twice-differentiable convex function ($f''(x) \ge 0$). Prove Jensen's Inequality for $n$ points $x_1, \dots, x_n \in I$ with non-negative weights $w_i \ge 0$ summing to $\sum_{i=1}^n w_i = 1$:

$$
f\left( \sum_{i=1}^n w_i x_i \right) \le \sum_{i=1}^n w_i f(x_i)
$$

#### Intuition
A convex function always lies **above** its tangent line. Writing the linear tangent approximation around the weighted mean $x_0 = \sum w_i x_i$ proves Jensen's inequality directly.

#### Solution
1. **Define Weighted Mean Center:**  
   Let $x_0 = \sum_{i=1}^n w_i x_i \in I$.

2. **Construct Tangent Line Inequality via Taylor's Theorem:**  
   For any point $x \in I$, expanding $f(x)$ around $x_0$:

$$
   f(x) = f(x_0) + f'(x_0)(x - x_0) + \frac{f''(\xi)}{2}(x - x_0)^2
$$

   Since $f''(x) \ge 0$ for all $x \in I$, the remainder term $\frac{f''(\xi)}{2}(x - x_0)^2 \ge 0$.  
   Therefore, for all $x \in I$:

$$
   f(x) \ge f(x_0) + f'(x_0)(x - x_0)
$$

3. **Evaluate at each $x_i$ and Multiply by $w_i$:**  

$$
   w_i f(x_i) \ge w_i f(x_0) + w_i f'(x_0)(x_i - x_0)
$$

4. **Sum over all $i = 1, \dots, n$:**  

$$
   \sum_{i=1}^n w_i f(x_i) \ge \sum_{i=1}^n w_i f(x_0) + f'(x_0) \sum_{i=1}^n w_i (x_i - x_0)
$$

5. **Simplify Sums:**  
   - $\sum_{i=1}^n w_i f(x_0) = f(x_0) \sum_{i=1}^n w_i = f(x_0)(1) = f(x_0)$.
   - $\sum_{i=1}^n w_i (x_i - x_0) = \sum w_i x_i - x_0 \sum w_i = x_0 - x_0(1) = 0$.

6. **Final Result:**  

$$
   \sum_{i=1}^n w_i f(x_i) \ge f(x_0) = f\left( \sum_{i=1}^n w_i x_i \right)
$$

$$
\boxed{f\left(\sum_{i=1}^n w_i x_i\right) \le \sum_{i=1}^n w_i f(x_i)}
$$

#### Key takeaway
Every differentiable convex function satisfies $f(x) \ge T_1(x)$; summing weighted tangent inequalities yields Jensen's Inequality in a single step.

**Verification.**

In [38]:
phi = lambda t: np.exp(t)          # strictly convex
pts = rng.uniform(-2.0, 2.0, size=6)
w = rng.dirichlet(np.ones(6))
lhs = phi(w @ pts)
rhs = w @ phi(pts)
print("weights       :", w)
print(f"f(sum w_i x_i) = {lhs:.12f}   <=   sum w_i f(x_i) = {rhs:.12f}")
print("Jensen holds  :", lhs <= rhs)
assert lhs <= rhs

weights       : [0.002  0.2266 0.2338 0.0835 0.0022 0.4519]
f(sum w_i x_i) = 2.831053411247   <=   sum w_i f(x_i) = 3.642159174321
Jensen holds  : True


The convex combination inequality holds for a random six-point weighting, as Jensen's inequality asserts.

### Problem L3.6 — Pathological L'Hôpital Counterexample Deep Dive
**Source:** Demidovich No. 1388 / Spivak Chapter 11  
**Statement.** Evaluate $\lim_{x \to 0} \frac{x^2 \sin(\frac{1}{x})}{\sin x}$. Show that L'Hôpital's Rule cannot be used because $\lim_{x \to 0} \frac{f'(x)}{g'(x)}$ does not exist, yet prove from first principles that the limit exists and equals $0$.

#### Intuition
Functions containing $x^k \sin(\frac{1}{x})$ exhibit wild oscillations near $x=0$, causing derivatives to oscillate wildly while the original function values are tightly bounded.

#### Solution
1. **Identify Indeterminate Form:**  
   As $x \to 0$, $\lvert x^2 \sin(\frac{1}{x}) \rvert \le x^2 \to 0$ and $\sin x \to 0$. Form is $\frac{0}{0}$.

2. **Compute Derivatives for L'Hôpital:**  

$$
   f(x) = x^2 \sin(\frac{1}{x}) \implies f'(x) = 2x \sin(\frac{1}{x}) + x^2 \cos(\frac{1}{x}) \cdot \left(-\frac{1}{x^2}\right) = 2x \sin(\frac{1}{x}) - \cos(\frac{1}{x})
$$

$$
   g(x) = \sin x \implies g'(x) = \cos x
$$

$$
   \frac{f'(x)}{g'(x)} = \frac{2x \sin(\frac{1}{x}) - \cos(\frac{1}{x})}{\cos x}
$$

3. **Examine Limit of Derivative Ratio:**  
   As $x \to 0$, $\cos x \to 1$ and $2x \sin(\frac{1}{x}) \to 0$.  
   However, $\cos(\frac{1}{x})$ oscillates endlessly between $-1$ and $+1$ as $x \to 0$.  
   Therefore, $\lim_{x \to 0} \frac{f'(x)}{g'(x)}$ **does not exist**. L'Hôpital's Rule provides zero information.

4. **First-Principles Direct Squeeze Theorem Evaluation:**  
   Rewrite the ratio:

$$
   \frac{x^2 \sin(\frac{1}{x})}{\sin x} = \left( \frac{x}{\sin x} \right) \cdot x \sin(\frac{1}{x})
$$

   - We know $\lim_{x \to 0} \frac{x}{\sin x} = 1$.
   - For $x \sin(\frac{1}{x})$: $-\lvert x \rvert \le x \sin(\frac{1}{x}) \le \lvert x \rvert$. As $x \to 0$, $\lim_{x \to 0} x \sin(\frac{1}{x}) = 0$ by Squeeze Theorem.
   - Product of limits: $1 \cdot 0 = 0$.

$$
\boxed{\text{Limit } = 0\text{. L'Hôpital fails because } \lim \frac{f'}{g'} \text{ oscillates, but Squeeze Theorem proves limit is } 0.}
$$

#### Key takeaway
L'Hôpital's Rule requires the limit of derivative ratios to exist; when rapid oscillation invalidates $f'(x)$, bounding inequalities resolve the limit effortlessly.

**Verification.**

In [39]:
xs = np.array([1e-1, 1e-2, 1e-3, 1e-4, 1e-5])
print("f/g near 0 :", xs**2 * np.sin(1 / xs) / np.sin(xs))
ts = np.array([1 / (k * np.pi) for k in range(100, 110)])
print("f'/g' at 1/(k pi):", (2 * ts * np.sin(1 / ts) - np.cos(1 / ts)) / np.cos(ts))
print("the ratio of derivatives alternates near +-1 and has no limit; the quotient goes to 0")
assert abs(1e-5**2 * np.sin(1 / 1e-5) / np.sin(1e-5)) < 1e-4

f/g near 0 : [-0.0545 -0.0051  0.0008 -0.      0.    ]
f'/g' at 1/(k pi): [-1.  1. -1.  1. -1.  1. -1.  1. -1.  1.]
the ratio of derivatives alternates near +-1 and has no limit; the quotient goes to 0


The quotient collapses to $0$ while the ratio of derivatives keeps swinging near $\pm 1$.

### Problem L3.7 — Nocedal & Wright — Armijo Line Search Finite Termination
**Source:** Nocedal & Wright, *Numerical Optimization*, Chapter 3  
**Statement.** Let $f: \mathbb{R} \to \mathbb{R}$ be continuously differentiable with $L$-Lipschitz continuous derivative ($\lvert f'(x) - f'(y) \rvert \le L \lvert x - y \rvert$). Let $d = -f'(x_k)$ be the steepest descent direction with $f'(x_k) \ne 0$. Prove that the Armijo condition $f(x_k + \alpha d) \le f(x_k) + c_1 \alpha f'(x_k) d$ ($c_1 \in (0, 1)$) is satisfied for all step sizes $\alpha \in \left[0, \frac{2(1-c_1)}{L}\right]$.

#### Intuition
Armijo line search requires that the function value decreases at least proportionally to step size $\alpha$. Lipschitz continuity of $f'$ bounds how fast the function can curve upward, guaranteeing a range of step sizes that satisfy sufficient decrease.

#### Solution
1. **Apply Fundamental Theorem of Calculus:**  

$$
   f(x_k + \alpha d) - f(x_k) = \int_0^1 f'(x_k + t \alpha d) \cdot (\alpha d) \, dt
$$

$$
   = \alpha f'(x_k) d + \int_0^1 \left[ f'(x_k + t \alpha d) - f'(x_k) \right] \alpha d \, dt
$$

2. **Bound Remainder using Lipschitz Continuity:**  
   Since $\lvert f'(x_k + t \alpha d) - f'(x_k) \rvert \le L t \alpha \lvert d \rvert$:

$$
   f(x_k + \alpha d) - f(x_k) \le \alpha f'(x_k) d + \int_0^1 L t \alpha^2 d^2 \, dt = \alpha f'(x_k) d + \frac{1}{2} L \alpha^2 d^2
$$

3. **Substitute Descent Direction $d = -f'(x_k)$:**  
   Note $f'(x_k) d = -\lvert f'(x_k) \rvert^2$ and $d^2 = \lvert f'(x_k) \rvert^2$.  

$$
   f(x_k + \alpha d) - f(x_k) \le -\alpha \lvert f'(x_k) \rvert^2 + \frac{1}{2} L \alpha^2 \lvert f'(x_k) \rvert^2 = -\alpha \lvert f'(x_k) \rvert^2 \left( 1 - \frac{1}{2} L \alpha \right)
$$

4. **Set Upper Bound $\le$ Armijo Target:**  
   We want:

$$
   -\alpha \lvert f'(x_k) \rvert^2 \left( 1 - \frac{1}{2} L \alpha \right) \le -c_1 \alpha \lvert f'(x_k) \rvert^2
$$

   Divide by $-\alpha \lvert f'(x_k) \rvert^2 \lt 0$ (reversing inequality):

$$
   1 - \frac{1}{2} L \alpha \ge c_1 \implies \frac{1}{2} L \alpha \le 1 - c_1 \implies \alpha \le \frac{2(1 - c_1)}{L}
$$

$$
\boxed{\text{Armijo condition holds for all } \alpha \in \left(0, \frac{2(1-c_1)}{L}\right]}
$$

#### Key takeaway
The Lipschitz constant $L$ of gradients dictates the maximum allowable step size $\alpha \le \frac{2(1-c_1)}{L}$ to guarantee monotonic decrease in line search optimization.

**Verification.**

In [40]:
L_const, c1 = 4.0, 0.3
f = lambda t: 0.5 * L_const * t**2
xk = 1.0
fp = L_const * xk
d = -fp
alpha_bound = 2 * (1 - c1) / L_const
grid = np.linspace(1e-4, 1.5 * alpha_bound, 20000)
ok = f(xk + grid * d) <= f(xk) + c1 * grid * fp * d
largest_ok = grid[ok].max()
print(f"predicted Armijo threshold 2(1-c1)/L = {alpha_bound:.10f}")
print(f"largest step passing the test        = {largest_ok:.10f}")
assert abs(largest_ok - alpha_bound) < 2e-4

predicted Armijo threshold 2(1-c1)/L = 0.3500000000
largest step passing the test        = 0.3499895895


The largest step passing the Armijo test coincides with the predicted threshold $2(1-c_1)/L$.

### Problem L3.8 — Putnam 1998 B1 — Global Extrema of High-Degree Factorization
**Source:** Stewart, *Calculus: Early Transcendentals*, Ch. 4 (standard constrained optimization exercise).  
**Statement.** Let $a, b$ be positive integers. Find the global maximum value of $f(x) = (x+1)^a (1-x)^b$ on the closed domain $[-1, 1]$, and prove uniqueness of the maximizing point $x^{\ast}$.

#### Intuition
The function is non-negative on $[-1, 1]$ and equals zero at both endpoints $x = -1$ and $x = 1$. By Rolle's / EVT, it must achieve a positive maximum at an interior critical point $x^{\ast} \in (-1, 1)$. Logarithmic differentiation converts high-degree exponent multiplication into a simple linear equation.

#### Solution
1. **Analyze Endpoint Values:**  
   $f(-1) = 0^a \cdot 2^b = 0$.  
   $f(1) = 2^a \cdot 0^b = 0$.  
   For $x \in (-1, 1)$, $(x+1) \gt 0$ and $(1-x) \gt 0$, so $f(x) \gt 0$.

2. **Logarithmic Differentiation:**  
   Take natural log for $x \in (-1, 1)$:

$$
   \ln f(x) = a \ln(x+1) + b \ln(1-x)
$$

   Differentiate with respect to $x$:

$$
   \frac{f'(x)}{f(x)} = \frac{a}{x+1} - \frac{b}{1-x} = \frac{a(1-x) - b(x+1)}{(x+1)(1-x)}
$$

3. **Find Interior Critical Point $x^{\ast}$:**  
   Set numerator $a(1-x) - b(x+1) = 0$:

$$
   a - ax - bx - b = 0 \implies (a+b)x = a - b \implies x^{\ast} = \frac{a - b}{a + b}
$$

4. **Verify Location of $x^{\ast}$:**  
   Since $a, b \gt 0$, $- (a+b) \lt a - b \lt a + b$, so $x^{\ast} = \frac{a-b}{a+b} \in (-1, 1)$.

5. **Uniqueness & Strict Maximum:**  
   $\frac{f'(x)}{f(x)} = \frac{a-b - (a+b)x}{1-x^2}$.
   - For $-1 \lt x \lt x^{\ast}$: numerator $\gt 0 \implies f'(x) \gt 0$ (increasing).
   - For $x^{\ast} \lt x \lt 1$: numerator $\lt 0 \implies f'(x) \lt 0$ (decreasing).  
   Hence $x^{\ast}$ is the unique strict global maximum.

6. **Calculate Maximum Value $f(x^{\ast})$:**  

$$
   x^{\ast} + 1 = \frac{a-b}{a+b} + 1 = \frac{2a}{a+b}
$$

$$
   1 - x^{\ast} = 1 - \frac{a-b}{a+b} = \frac{2b}{a+b}
$$

$$
   f(x^{\ast}) = \left( \frac{2a}{a+b} \right)^a \left( \frac{2b}{a+b} \right)^b = \frac{2^{a+b} a^a b^b}{(a+b)^{a+b}}
$$

$$
\boxed{x^{\ast} = \frac{a-b}{a+b}, \quad f_{\text{max}} = \frac{2^{a+b} a^a b^b}{(a+b)^{a+b}}}
$$

#### Key takeaway
Logarithmic differentiation simplifies high-power binomial products into linear root equations for instantaneous global maximum determination.

**Verification.**

In [41]:
a_e, b_e = 3.0, 5.0
f = lambda t: (1 + t) ** a_e * (1 - t) ** b_e
res = optimize.minimize_scalar(lambda t: -f(t), bounds=(-1.0, 1.0), method="bounded",
                               options={"xatol": 1e-13})
x_star = (a_e - b_e) / (a_e + b_e)
f_star = 2 ** (a_e + b_e) * a_e**a_e * b_e**b_e / (a_e + b_e) ** (a_e + b_e)
print(f"numerical argmax = {res.x:.10f}   formula (a-b)/(a+b) = {x_star:.10f}")
print(f"numerical max    = {f(res.x):.10f}   formula             = {f_star:.10f}")
assert abs(res.x - x_star) < 1e-6 and abs(f(res.x) - f_star) < 1e-8

numerical argmax = -0.2499999963   formula (a-b)/(a+b) = -0.2500000000
numerical max    = 1.2874603271   formula             = 1.2874603271


The numerical maximizer and maximum agree with the closed-form Putnam answer to eight digits.